// fintoc/build/module/lib/client.js
import crypto from 'crypto';
import axios from 'axios';
import { JWSSignature } from './jws';
import { paginate } from './paginator';
export class Client {
    static #client;
    baseUrl;
    apiKey;
    userAgent;
    params;
    jws;
    constructor(options) {
        const { baseUrl, apiKey, userAgent, params = {}, jwsPrivateKey, } = options;
        this.baseUrl = baseUrl;
        this.apiKey = apiKey;
        this.userAgent = userAgent;
        this.params = params || {};
        if (jwsPrivateKey) {
            this.jws = new JWSSignature(jwsPrivateKey);
        }
    }
    get client() {
        if (Client.#client === undefined) {
            Client.#client = axios.create({
                baseURL: this.baseUrl,
                headers: this.staticHeaders,
                params: this.params,
            });
        }
        return Client.#client;
    }
    get staticHeaders() {
        return { Authorization: this.apiKey, 'User-Agent': this.userAgent, 'Content-Type': 'application/json' };
    }
    async request(options) {
        const { path, json, paginated = false, method = 'get', params = {}, idempotencyKey, } = options;
        const url = `${this.baseUrl}${path}`;
        const allParams = { ...this.params, ...params };
        if (paginated) {
            return paginate({
                client: this.client,
                path: url,
                headers: this.staticHeaders,
                params: allParams,
            });
        }
        const rawBody = JSON.stringify(json);
        const headers = this.buildHeaders(method, rawBody, idempotencyKey);
        const response = await this.client.request({
            method, url, params: allParams, data: rawBody, headers,
        });
        return response.data;
    }
    extend(extension) {
        const { baseUrl, apiKey, userAgent, params, } = (extension || {});
        return new Client({
            baseUrl: baseUrl || this.baseUrl,
            apiKey: apiKey || this.apiKey,
            userAgent: userAgent || this.userAgent,
            params: params ? { ...this.params, ...params } : { ...this.params },
        });
    }
    buildHeaders(method, rawBody, idempotencyKey) {
        const headers = {
            ...this.staticHeaders,
            ...this.buildJWSHeaders(method, rawBody),
        };
        if (method === 'post') {
            headers['Idempotency-Key'] = idempotencyKey || crypto.randomUUID();
        }
        return headers;
    }
    buildJWSHeaders(method, rawBody) {
        if (this.jws && (method === 'post' || method === 'put' || method === 'patch')) {
            const signature = this.jws.generateHeader(rawBody);
            return { 'Fintoc-JWS-Signature': signature };
        }
        return {};
    }
}
