// fintoc/build/main/spec/integration.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ava_1 = __importDefault(require("ava"));
const core_1 = require("../lib/core");
const initializers_1 = require("./mocks/initializers");
ava_1.default.before((t) => {
    const ctx = t.context;
    ctx.axiosMock = (0, initializers_1.mockAxios)();
});
ava_1.default.after((t) => {
    const ctx = t.context;
    (0, initializers_1.restore)(ctx.axiosMock);
});
ava_1.default.beforeEach((t) => {
    const ctx = t.context;
    const apiKey = 'super_secret_api_key';
    ctx.fintoc = new core_1.Fintoc(apiKey);
});
(0, ava_1.default)('fintoc.paymentIntents.list()', async (t) => {
    const ctx = t.context;
    const paymentIntents = await ctx.fintoc.paymentIntents.list();
    let count = 0;
    for await (const paymentIntent of paymentIntents) {
        count += 1;
        t.is(paymentIntent.method, 'get');
        t.is(paymentIntent.url, 'v1/payment_intents');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.paymentIntents.all() - deprecated', async (t) => {
    const ctx = t.context;
    const paymentIntents = await ctx.fintoc.paymentIntents.all();
    let count = 0;
    for await (const paymentIntent of paymentIntents) {
        count += 1;
        t.is(paymentIntent.method, 'get');
        t.is(paymentIntent.url, 'v1/payment_intents');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.paymentIntents.get()', async (t) => {
    const ctx = t.context;
    const paymentId = 'payment_id';
    const paymentIntent = await ctx.fintoc.paymentIntents.get(paymentId);
    t.is(paymentIntent.method, 'get');
    t.is(paymentIntent.url, `v1/payment_intents/${paymentId}`);
});
(0, ava_1.default)('fintoc.paymentIntents.create()', async (t) => {
    const ctx = t.context;
    const paymentData = {
        amount: 1000,
        currency: 'USD',
        payment_type: 'bank_transfer',
    };
    const paymentIntent = await ctx.fintoc.paymentIntents.create(paymentData);
    t.is(paymentIntent.method, 'post');
    t.is(paymentIntent.url, 'v1/payment_intents');
    t.is(paymentIntent.json.amount, paymentData.amount);
    t.is(paymentIntent.json.currency, paymentData.currency);
    t.is(paymentIntent.json.payment_type, paymentData.payment_type);
});
(0, ava_1.default)('fintoc.paymentIntents.expire()', async (t) => {
    const ctx = t.context;
    const paymentIntentId = 'pi_2x9Q2ufili2A5Jw3ugI3CrdDHtS';
    const expiredPaymentIntent = await ctx.fintoc.paymentIntents.expire(paymentIntentId);
    t.is(expiredPaymentIntent.method, 'post');
    t.is(expiredPaymentIntent.url, `v1/payment_intents/${paymentIntentId}/expire`);
});
(0, ava_1.default)('fintoc.paymentIntents.checkEligibility()', async (t) => {
    const ctx = t.context;
    const eligibilityData = {
        amount: 10000,
        currency: 'CLP',
    };
    const result = await ctx.fintoc.paymentIntents.checkEligibility(eligibilityData);
    t.is(result.method, 'post');
    t.is(result.url, 'v1/payment_intents/check_eligibility');
});
(0, ava_1.default)('fintoc.paymentLinks.list()', async (t) => {
    const ctx = t.context;
    const paymentLinks = await ctx.fintoc.paymentLinks.list();
    let count = 0;
    for await (const paymentLink of paymentLinks) {
        count += 1;
        t.is(paymentLink.method, 'get');
        t.is(paymentLink.url, 'v1/payment_links');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.paymentLinks.get()', async (t) => {
    const ctx = t.context;
    const paymentLinkId = 'pl_example_id';
    const paymentLink = await ctx.fintoc.paymentLinks.get(paymentLinkId);
    t.is(paymentLink.method, 'get');
    t.is(paymentLink.url, `v1/payment_links/${paymentLinkId}`);
});
(0, ava_1.default)('fintoc.paymentLinks.create()', async (t) => {
    const ctx = t.context;
    const paymentLinkData = {
        amount: 5000,
        currency: 'CLP',
        description: 'Payment for services',
    };
    const paymentLink = await ctx.fintoc.paymentLinks.create(paymentLinkData);
    t.is(paymentLink.method, 'post');
    t.is(paymentLink.url, 'v1/payment_links');
    t.is(paymentLink.json.amount, paymentLinkData.amount);
    t.is(paymentLink.json.currency, paymentLinkData.currency);
    t.is(paymentLink.json.description, paymentLinkData.description);
});
(0, ava_1.default)('fintoc.paymentLinks.cancel()', async (t) => {
    const ctx = t.context;
    const paymentLinkId = 'pl_example_id';
    const cancelledPaymentLink = await ctx.fintoc.paymentLinks.cancel(paymentLinkId);
    t.is(cancelledPaymentLink.method, 'patch');
    t.is(cancelledPaymentLink.url, `v1/payment_links/${paymentLinkId}/cancel`);
});
(0, ava_1.default)('fintoc.links.list()', async (t) => {
    const ctx = t.context;
    const links = await ctx.fintoc.links.list();
    let count = 0;
    for await (const link of links) {
        count += 1;
        t.is(link.method, 'get');
        t.is(link.url, 'v1/links');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.links.all() - deprecated', async (t) => {
    const ctx = t.context;
    const links = await ctx.fintoc.links.all();
    let count = 0;
    for await (const link of links) {
        count += 1;
        t.is(link.method, 'get');
        t.is(link.url, 'v1/links');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.links.get()', async (t) => {
    const ctx = t.context;
    const linkToken = 'link_token_example';
    const link = await ctx.fintoc.links.get(linkToken);
    t.is(link.method, 'get');
    t.is(link.url, `v1/links/${linkToken}`);
});
(0, ava_1.default)('fintoc.links.update()', async (t) => {
    const ctx = t.context;
    const linkToken = 'link_token_example';
    const updateData = {
        username: 'updated_username',
    };
    const link = await ctx.fintoc.links.update(linkToken, updateData);
    t.is(link.method, 'patch');
    t.is(link.url, `v1/links/${linkToken}`);
    t.is(link.json.username, updateData.username);
});
(0, ava_1.default)('fintoc.links.delete()', async (t) => {
    const ctx = t.context;
    const linkId = 'link_id';
    const deletedIdentifier = await ctx.fintoc.links.delete(linkId);
    t.is(deletedIdentifier, linkId);
});
(0, ava_1.default)('fintoc.accounts.list()', async (t) => {
    const ctx = t.context;
    const accounts = await ctx.fintoc.accounts.list();
    let count = 0;
    for await (const account of accounts) {
        count += 1;
        t.is(account.method, 'get');
        t.is(account.url, 'v1/accounts');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.accounts.all() - deprecated', async (t) => {
    const ctx = t.context;
    const accounts = await ctx.fintoc.accounts.all();
    let count = 0;
    for await (const account of accounts) {
        count += 1;
        t.is(account.method, 'get');
        t.is(account.url, 'v1/accounts');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.accounts.get()', async (t) => {
    const ctx = t.context;
    const accountId = 'account_id';
    const account = await ctx.fintoc.accounts.get(accountId);
    t.is(account.method, 'get');
    t.is(account.url, `v1/accounts/${accountId}`);
});
(0, ava_1.default)('fintoc.accounts.movements.list()', async (t) => {
    const ctx = t.context;
    const linkToken = 'link_token_example';
    const accountId = 'account_id';
    const movements = await ctx.fintoc.accounts.movements.list({
        account_id: accountId,
        link_token: linkToken,
    });
    let count = 0;
    for await (const movement of movements) {
        count += 1;
        t.is(movement.method, 'get');
        t.is(movement.url, `v1/accounts/${accountId}/movements`);
        t.is(movement.params.link_token, linkToken);
        t.is(movement.params.account_id, accountId);
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.accounts.movements.all() - deprecated', async (t) => {
    const ctx = t.context;
    const linkToken = 'link_token_example';
    const accountId = 'account_id';
    const movements = await ctx.fintoc.accounts.movements.all({
        account_id: accountId,
        link_token: linkToken,
    });
    let count = 0;
    for await (const movement of movements) {
        count += 1;
        t.is(movement.method, 'get');
        t.is(movement.url, `v1/accounts/${accountId}/movements`);
        t.is(movement.params.link_token, linkToken);
        t.is(movement.params.account_id, accountId);
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.accounts.movements.get()', async (t) => {
    const ctx = t.context;
    const linkToken = 'link_token_example';
    const accountId = 'account_id';
    const movementId = 'movement_id';
    const movement = await ctx.fintoc.accounts.movements.get(movementId, {
        account_id: accountId,
        link_token: linkToken,
    });
    t.is(movement.method, 'get');
    t.is(movement.url, `v1/accounts/${accountId}/movements/${movementId}`);
    t.is(movement.params.link_token, linkToken);
    t.is(movement.params.account_id, accountId);
});
(0, ava_1.default)('fintoc.webhookEndpoints.list()', async (t) => {
    const ctx = t.context;
    const webhookEndpoints = await ctx.fintoc.webhookEndpoints.list();
    let count = 0;
    for await (const webhookEndpoint of webhookEndpoints) {
        count += 1;
        t.is(webhookEndpoint.method, 'get');
        t.is(webhookEndpoint.url, 'v1/webhook_endpoints');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.webhookEndpoints.all() - deprecated', async (t) => {
    const ctx = t.context;
    const webhookEndpoints = await ctx.fintoc.webhookEndpoints.all();
    let count = 0;
    for await (const webhookEndpoint of webhookEndpoints) {
        count += 1;
        t.is(webhookEndpoint.method, 'get');
        t.is(webhookEndpoint.url, 'v1/webhook_endpoints');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.webhookEndpoints.get()', async (t) => {
    const ctx = t.context;
    const webhookEndpointId = 'we_example_id';
    const webhookEndpoint = await ctx.fintoc.webhookEndpoints.get(webhookEndpointId);
    t.is(webhookEndpoint.method, 'get');
    t.is(webhookEndpoint.url, `v1/webhook_endpoints/${webhookEndpointId}`);
});
(0, ava_1.default)('fintoc.webhookEndpoints.create()', async (t) => {
    const ctx = t.context;
    const webhookData = {
        url: 'https://example.com/webhook',
        enabled_events: ['payment_intent.succeeded'],
    };
    const webhookEndpoint = await ctx.fintoc.webhookEndpoints.create(webhookData);
    t.is(webhookEndpoint.method, 'post');
    t.is(webhookEndpoint.url, 'v1/webhook_endpoints');
    t.is(webhookEndpoint.json.url, webhookData.url);
    t.deepEqual(webhookEndpoint.json.enabled_events, webhookData.enabled_events);
});
(0, ava_1.default)('fintoc.webhookEndpoints.update()', async (t) => {
    const ctx = t.context;
    const webhookEndpointId = 'we_example_id';
    const updateData = {
        enabled_events: ['payment_intent.failed', 'refund.succeeded'],
    };
    const webhookEndpoint = await ctx.fintoc.webhookEndpoints.update(webhookEndpointId, updateData);
    t.is(webhookEndpoint.method, 'patch');
    t.is(webhookEndpoint.url, `v1/webhook_endpoints/${webhookEndpointId}`);
    t.deepEqual(webhookEndpoint.json.enabled_events, updateData.enabled_events);
});
(0, ava_1.default)('fintoc.webhookEndpoints.delete()', async (t) => {
    const ctx = t.context;
    const webhookEndpointId = 'we_example_id';
    const deletedIdentifier = await ctx.fintoc.webhookEndpoints.delete(webhookEndpointId);
    t.is(deletedIdentifier, webhookEndpointId);
});
(0, ava_1.default)('fintoc.invoices.list()', async (t) => {
    const ctx = t.context;
    const invoices = await ctx.fintoc.invoices.list();
    let count = 0;
    for await (const invoice of invoices) {
        count += 1;
        t.is(invoice.method, 'get');
        t.is(invoice.url, 'v1/invoices');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.invoices.all() - deprecated', async (t) => {
    const ctx = t.context;
    const invoices = await ctx.fintoc.invoices.all();
    let count = 0;
    for await (const invoice of invoices) {
        count += 1;
        t.is(invoice.method, 'get');
        t.is(invoice.url, 'v1/invoices');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.refreshIntents.list()', async (t) => {
    const ctx = t.context;
    const refreshIntents = await ctx.fintoc.refreshIntents.list();
    let count = 0;
    for await (const refreshIntent of refreshIntents) {
        count += 1;
        t.is(refreshIntent.method, 'get');
        t.is(refreshIntent.url, 'v1/refresh_intents');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.refreshIntents.all() - deprecated', async (t) => {
    const ctx = t.context;
    const refreshIntents = await ctx.fintoc.refreshIntents.all();
    let count = 0;
    for await (const refreshIntent of refreshIntents) {
        count += 1;
        t.is(refreshIntent.method, 'get');
        t.is(refreshIntent.url, 'v1/refresh_intents');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.refreshIntents.get()', async (t) => {
    const ctx = t.context;
    const refreshIntentId = 'refresh_intent_id';
    const refreshIntent = await ctx.fintoc.refreshIntents.get(refreshIntentId);
    t.is(refreshIntent.method, 'get');
    t.is(refreshIntent.url, `v1/refresh_intents/${refreshIntentId}`);
});
(0, ava_1.default)('fintoc.refreshIntents.create()', async (t) => {
    const ctx = t.context;
    const refreshData = {
        link_token: 'link_token_example',
        refresh_type: 'only_last',
    };
    const refreshIntent = await ctx.fintoc.refreshIntents.create(refreshData);
    t.is(refreshIntent.method, 'post');
    t.is(refreshIntent.url, 'v1/refresh_intents');
    t.is(refreshIntent.json.link_token, refreshData.link_token);
    t.is(refreshIntent.json.refresh_type, refreshData.refresh_type);
});
(0, ava_1.default)('fintoc.subscriptions.list()', async (t) => {
    const ctx = t.context;
    const subscriptions = await ctx.fintoc.subscriptions.list();
    let count = 0;
    for await (const subscription of subscriptions) {
        count += 1;
        t.is(subscription.method, 'get');
        t.is(subscription.url, 'v1/subscriptions');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.subscriptions.all() - deprecated', async (t) => {
    const ctx = t.context;
    const subscriptions = await ctx.fintoc.subscriptions.all();
    let count = 0;
    for await (const subscription of subscriptions) {
        count += 1;
        t.is(subscription.method, 'get');
        t.is(subscription.url, 'v1/subscriptions');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.subscriptions.get()', async (t) => {
    const ctx = t.context;
    const subscriptionId = 'subscription_id';
    const subscription = await ctx.fintoc.subscriptions.get(subscriptionId);
    t.is(subscription.method, 'get');
    t.is(subscription.url, `v1/subscriptions/${subscriptionId}`);
});
(0, ava_1.default)('fintoc.taxReturns.list()', async (t) => {
    const ctx = t.context;
    const taxReturns = await ctx.fintoc.taxReturns.list();
    let count = 0;
    for await (const taxReturn of taxReturns) {
        count += 1;
        t.is(taxReturn.method, 'get');
        t.is(taxReturn.url, 'v1/tax_returns');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.taxReturns.all() - deprecated', async (t) => {
    const ctx = t.context;
    const taxReturns = await ctx.fintoc.taxReturns.all();
    let count = 0;
    for await (const taxReturn of taxReturns) {
        count += 1;
        t.is(taxReturn.method, 'get');
        t.is(taxReturn.url, 'v1/tax_returns');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.taxReturns.get()', async (t) => {
    const ctx = t.context;
    const taxReturnId = 'tax_return_id';
    const taxReturn = await ctx.fintoc.taxReturns.get(taxReturnId);
    t.is(taxReturn.method, 'get');
    t.is(taxReturn.url, `v1/tax_returns/${taxReturnId}`);
});
(0, ava_1.default)('fintoc.charges.list()', async (t) => {
    const ctx = t.context;
    const charges = await ctx.fintoc.charges.list();
    let count = 0;
    for await (const charge of charges) {
        count += 1;
        t.is(charge.method, 'get');
        t.is(charge.url, 'v1/charges');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.charges.get()', async (t) => {
    const ctx = t.context;
    const chargeId = 'charge_id';
    const charge = await ctx.fintoc.charges.get(chargeId);
    t.is(charge.method, 'get');
    t.is(charge.url, `v1/charges/${chargeId}`);
});
(0, ava_1.default)('fintoc.charges.create()', async (t) => {
    const ctx = t.context;
    const chargeData = {
        amount: 1000,
        currency: 'CLP',
        payment_method: 'bank_transfer',
    };
    const charge = await ctx.fintoc.charges.create(chargeData);
    t.is(charge.method, 'post');
    t.is(charge.url, 'v1/charges');
    t.is(charge.json.amount, chargeData.amount);
    t.is(charge.json.currency, chargeData.currency);
    t.is(charge.json.payment_method, chargeData.payment_method);
});
(0, ava_1.default)('fintoc.refunds.list()', async (t) => {
    const ctx = t.context;
    const refunds = await ctx.fintoc.refunds.list();
    let count = 0;
    for await (const refund of refunds) {
        count += 1;
        t.is(refund.method, 'get');
        t.is(refund.url, 'v1/refunds');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.refunds.get()', async (t) => {
    const ctx = t.context;
    const refundId = 'ref_QmbpWzP1HOngN3X7';
    const refund = await ctx.fintoc.refunds.get(refundId);
    t.is(refund.method, 'get');
    t.is(refund.url, `v1/refunds/${refundId}`);
});
(0, ava_1.default)('fintoc.refunds.create()', async (t) => {
    const ctx = t.context;
    const refundData = {
        resource_type: 'payment_intent',
        resource_id: 'pi_30yWq311fOLrAAKkSH1bvODVLGa',
        amount: 1000,
    };
    const refund = await ctx.fintoc.refunds.create(refundData);
    t.is(refund.method, 'post');
    t.is(refund.url, 'v1/refunds');
    t.is(refund.json.resource_type, refundData.resource_type);
    t.is(refund.json.resource_id, refundData.resource_id);
    t.is(refund.json.amount, refundData.amount);
});
(0, ava_1.default)('fintoc.refunds.cancel()', async (t) => {
    const ctx = t.context;
    const refundId = 'ref_QmbpWzP1HOngN3X7';
    const refund = await ctx.fintoc.refunds.cancel(refundId);
    t.is(refund.method, 'post');
    t.is(refund.url, `v1/refunds/${refundId}/cancel`);
});
(0, ava_1.default)('fintoc.subscriptionIntents.list()', async (t) => {
    const ctx = t.context;
    const subscriptionIntents = await ctx.fintoc.subscriptionIntents.list();
    let count = 0;
    for await (const subscriptionIntent of subscriptionIntents) {
        count += 1;
        t.is(subscriptionIntent.method, 'get');
        t.is(subscriptionIntent.url, 'v1/subscription_intents');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.subscriptionIntents.get()', async (t) => {
    const ctx = t.context;
    const subscriptionIntentId = 'subscription_intent_id';
    const subscriptionIntent = await ctx.fintoc.subscriptionIntents.get(subscriptionIntentId);
    t.is(subscriptionIntent.method, 'get');
    t.is(subscriptionIntent.url, `v1/subscription_intents/${subscriptionIntentId}`);
});
(0, ava_1.default)('fintoc.subscriptionIntents.create()', async (t) => {
    const ctx = t.context;
    const subscriptionIntentData = {
        amount: 1000,
        currency: 'CLP',
    };
    const subscriptionIntent = await ctx.fintoc.subscriptionIntents.create(subscriptionIntentData);
    t.is(subscriptionIntent.method, 'post');
    t.is(subscriptionIntent.url, 'v1/subscription_intents');
    t.is(subscriptionIntent.json.amount, subscriptionIntentData.amount);
    t.is(subscriptionIntent.json.currency, subscriptionIntentData.currency);
});
(0, ava_1.default)('fintoc.checkoutSessions.create()', async (t) => {
    const ctx = t.context;
    const checkoutSessionData = {
        amount: 5000,
        currency: 'CLP',
        success_url: 'https://example.com/success',
        cancel_url: 'https://example.com/cancel',
    };
    const checkoutSession = await ctx.fintoc.checkoutSessions.create(checkoutSessionData);
    t.is(checkoutSession.method, 'post');
    t.is(checkoutSession.url, 'v1/checkout_sessions');
    t.is(checkoutSession.json.amount, checkoutSessionData.amount);
    t.is(checkoutSession.json.currency, checkoutSessionData.currency);
    t.is(checkoutSession.json.success_url, checkoutSessionData.success_url);
    t.is(checkoutSession.json.cancel_url, checkoutSessionData.cancel_url);
});
(0, ava_1.default)('fintoc.checkoutSessions.get()', async (t) => {
    const ctx = t.context;
    const checkoutSessionId = 'checkout_session_id';
    const checkoutSession = await ctx.fintoc.checkoutSessions.get(checkoutSessionId);
    t.is(checkoutSession.method, 'get');
    t.is(checkoutSession.url, `v1/checkout_sessions/${checkoutSessionId}`);
});
(0, ava_1.default)('fintoc.checkoutSessions.expire()', async (t) => {
    const ctx = t.context;
    const checkoutSessionId = 'checkout_session_id';
    const checkoutSession = await ctx.fintoc.checkoutSessions.expire(checkoutSessionId);
    t.is(checkoutSession.method, 'post');
    t.is(checkoutSession.url, `v1/checkout_sessions/${checkoutSessionId}/expire`);
});
(0, ava_1.default)('link.accounts.list()', async (t) => {
    const ctx = t.context;
    const linkToken = 'link_token_example';
    const link = await ctx.fintoc.links.get(linkToken);
    const accounts = await link.accounts.list();
    t.is(link.accounts._client.params.link_token, linkToken);
    let count = 0;
    for await (const account of accounts) {
        count += 1;
        t.is(account.method, 'get');
        t.is(account.url, 'v1/accounts');
    }
    t.true(count > 0);
});
(0, ava_1.default)('link.accounts.all() - deprecated', async (t) => {
    const ctx = t.context;
    const linkToken = 'link_token_example';
    const link = await ctx.fintoc.links.get(linkToken);
    const accounts = await link.accounts.all();
    t.is(link.accounts._client.params.link_token, linkToken);
    let count = 0;
    for await (const account of accounts) {
        count += 1;
        t.is(account.method, 'get');
        t.is(account.url, 'v1/accounts');
    }
    t.true(count > 0);
});
(0, ava_1.default)('link.accounts.get()', async (t) => {
    const ctx = t.context;
    const linkToken = 'link_token_example';
    const link = await ctx.fintoc.links.get(linkToken);
    const accountId = 'account_id';
    const account = await link.accounts.get(accountId);
    t.is(account.method, 'get');
    t.is(account.url, `v1/accounts/${accountId}`);
    t.is(link.accounts._client.params.link_token, linkToken);
    t.is(account._client.params.link_token, linkToken);
});
(0, ava_1.default)('account.movements.list()', async (t) => {
    const ctx = t.context;
    const linkToken = 'link_token_example';
    const link = await ctx.fintoc.links.get(linkToken);
    const accountId = 'account_id';
    const account = await link.accounts.get(accountId);
    const movements = await account.movements.list();
    t.is(account.movements._client.params.link_token, linkToken);
    let count = 0;
    for await (const movement of movements) {
        count += 1;
        t.is(movement.method, 'get');
        t.is(movement.url, `v1/accounts/${accountId}/movements`);
        t.is(movement._client.params.link_token, linkToken);
    }
    t.true(count > 0);
});
(0, ava_1.default)('account.movements.all() - deprecated', async (t) => {
    const ctx = t.context;
    const linkToken = 'link_token_example';
    const link = await ctx.fintoc.links.get(linkToken);
    const accountId = 'account_id';
    const account = await link.accounts.get(accountId);
    const movements = await account.movements.all();
    t.is(account.movements._client.params.link_token, linkToken);
    let count = 0;
    for await (const movement of movements) {
        count += 1;
        t.is(movement.method, 'get');
        t.is(movement.url, `v1/accounts/${accountId}/movements`);
        t.is(movement._client.params.link_token, linkToken);
    }
    t.true(count > 0);
});
(0, ava_1.default)('account.movements.get()', async (t) => {
    const ctx = t.context;
    const linkToken = 'link_token_example';
    const link = await ctx.fintoc.links.get(linkToken);
    const accountId = 'account_id';
    const account = await link.accounts.get(accountId);
    const movementId = 'movement_id';
    const movement = await account.movements.get(movementId);
    t.is(movement.method, 'get');
    t.is(movement.url, `v1/accounts/${accountId}/movements/${movementId}`);
    t.is(movement._client.params.link_token, linkToken);
});
(0, ava_1.default)('fintoc.v2.transfers.list()', async (t) => {
    const ctx = t.context;
    const transfers = await ctx.fintoc.v2.transfers.list();
    let count = 0;
    for await (const transfer of transfers) {
        count += 1;
        t.is(transfer.method, 'get');
        t.is(transfer.url, 'v2/transfers');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.v2.transfers.get()', async (t) => {
    const ctx = t.context;
    const transferId = 'transfer_id';
    const transfer = await ctx.fintoc.v2.transfers.get(transferId);
    t.is(transfer.method, 'get');
    t.is(transfer.url, `v2/transfers/${transferId}`);
});
(0, ava_1.default)('fintoc.v2.transfers.create()', async (t) => {
    const ctx = t.context;
    const transferData = {
        amount: 10000,
        currency: 'MXN',
        counterparty: {
            account_number: '012969100000000026',
        },
    };
    const transfer = await ctx.fintoc.v2.transfers.create(transferData);
    t.is(transfer.method, 'post');
    t.is(transfer.url, 'v2/transfers');
    t.is(transfer.json.amount, transferData.amount);
    t.is(transfer.json.currency, transferData.currency);
    t.is(transfer.json.counterparty.account_number, transferData.counterparty.account_number);
    t.true(transfer.headers['Idempotency-Key'] && transfer.headers['Idempotency-Key'].length > 0);
});
(0, ava_1.default)('fintoc.v2.transfers.create() with idempotency key', async (t) => {
    const ctx = t.context;
    const transferData = {
        amount: 10000,
        currency: 'MXN',
        idempotency_key: 'idempotency_key_123',
        counterparty: {
            account_number: '012969100000000026',
        },
    };
    const transfer = await ctx.fintoc.v2.transfers.create(transferData);
    t.is(transfer.method, 'post');
    t.is(transfer.url, 'v2/transfers');
    t.is(transfer.headers['Idempotency-Key'], transferData.idempotency_key);
});
(0, ava_1.default)('fintoc.v2.accounts.list()', async (t) => {
    const ctx = t.context;
    const accounts = await ctx.fintoc.v2.accounts.list();
    let count = 0;
    for await (const account of accounts) {
        count += 1;
        t.is(account.method, 'get');
        t.is(account.url, 'v2/accounts');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.v2.accounts.get()', async (t) => {
    const ctx = t.context;
    const accountId = 'account_id';
    const account = await ctx.fintoc.v2.accounts.get(accountId);
    t.is(account.method, 'get');
    t.is(account.url, `v2/accounts/${accountId}`);
});
(0, ava_1.default)('fintoc.v2.accounts.create()', async (t) => {
    const ctx = t.context;
    const accountData = {
        description: 'Test Account',
    };
    const account = await ctx.fintoc.v2.accounts.create(accountData);
    t.is(account.method, 'post');
    t.is(account.url, 'v2/accounts');
    t.is(account.json.description, accountData.description);
});
(0, ava_1.default)('fintoc.v2.accountNumbers.list()', async (t) => {
    const ctx = t.context;
    const accountNumbers = await ctx.fintoc.v2.accountNumbers.list();
    let count = 0;
    for await (const accountNumber of accountNumbers) {
        count += 1;
        t.is(accountNumber.method, 'get');
        t.is(accountNumber.url, 'v2/account_numbers');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.v2.accountNumbers.get()', async (t) => {
    const ctx = t.context;
    const accountNumberId = 'account_number_id';
    const accountNumber = await ctx.fintoc.v2.accountNumbers.get(accountNumberId);
    t.is(accountNumber.method, 'get');
    t.is(accountNumber.url, `v2/account_numbers/${accountNumberId}`);
});
(0, ava_1.default)('fintoc.v2.accountNumbers.create()', async (t) => {
    const ctx = t.context;
    const accountNumberData = {
        account_id: 'account_123',
        description: 'test acc number',
    };
    const accountNumber = await ctx.fintoc.v2.accountNumbers.create(accountNumberData);
    t.is(accountNumber.method, 'post');
    t.is(accountNumber.url, 'v2/account_numbers');
    t.is(accountNumber.json.account_id, accountNumberData.account_id);
    t.is(accountNumber.json.description, accountNumberData.description);
});
(0, ava_1.default)('fintoc.v2.accountNumbers.update()', async (t) => {
    const ctx = t.context;
    const accountNumberId = 'account_number_id';
    const updateData = {
        status: 'enabled',
    };
    const accountNumber = await ctx.fintoc.v2.accountNumbers.update(accountNumberId, updateData);
    t.is(accountNumber.method, 'patch');
    t.is(accountNumber.url, `v2/account_numbers/${accountNumberId}`);
    t.is(accountNumber.json.status, updateData.status);
});
(0, ava_1.default)('fintoc.v2.transfers.return()', async (t) => {
    const ctx = t.context;
    const returnData = {
        transfer_id: 'transfer_123',
    };
    const result = await ctx.fintoc.v2.transfers.return(returnData);
    t.is(result.method, 'post');
    t.is(result.url, 'v2/transfers/return');
    t.is(result.json.transfer_id, returnData.transfer_id);
});
(0, ava_1.default)('fintoc.v2.simulate.receiveTransfer()', async (t) => {
    const ctx = t.context;
    const transferData = {
        amount: 50000,
        currency: 'MXN',
        account_number_id: 'account_number_123',
    };
    const result = await ctx.fintoc.v2.simulate.receiveTransfer(transferData);
    t.is(result.method, 'post');
    t.is(result.url, 'v2/simulate/receive_transfer');
    t.is(result.json.amount, transferData.amount);
    t.is(result.json.currency, transferData.currency);
    t.is(result.json.account_number_id, transferData.account_number_id);
});
(0, ava_1.default)('fintoc.v2.entities.list()', async (t) => {
    const ctx = t.context;
    const entities = await ctx.fintoc.v2.entities.list();
    let count = 0;
    for await (const entity of entities) {
        count += 1;
        t.is(entity.method, 'get');
        t.is(entity.url, 'v2/entities');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.v2.entities.get()', async (t) => {
    const ctx = t.context;
    const entityId = 'entity_id';
    const entity = await ctx.fintoc.v2.entities.get(entityId);
    t.is(entity.method, 'get');
    t.is(entity.url, `v2/entities/${entityId}`);
});
(0, ava_1.default)('fintoc.v2.accountVerifications.list()', async (t) => {
    const ctx = t.context;
    const accountVerifications = await ctx.fintoc.v2.accountVerifications.list();
    let count = 0;
    for await (const accountVerification of accountVerifications) {
        count += 1;
        t.is(accountVerification.method, 'get');
        t.is(accountVerification.url, 'v2/account_verifications');
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.v2.accountVerifications.get()', async (t) => {
    const ctx = t.context;
    const accountVerificationId = 'account_verification_id';
    const accountVerification = await ctx.fintoc.v2.accountVerifications.get(accountVerificationId);
    t.is(accountVerification.method, 'get');
    t.is(accountVerification.url, `v2/account_verifications/${accountVerificationId}`);
});
(0, ava_1.default)('fintoc.v2.accountVerifications.create()', async (t) => {
    const ctx = t.context;
    const accountVerificationData = {
        account_number: 'account_123',
    };
    const accountVerification = await ctx.fintoc.v2.accountVerifications.create(accountVerificationData);
    t.is(accountVerification.method, 'post');
    t.is(accountVerification.url, 'v2/account_verifications');
    t.is(accountVerification.json.account_number, accountVerificationData.account_number);
});
(0, ava_1.default)('fintoc.v2.accounts.movements.list()', async (t) => {
    const ctx = t.context;
    const accountId = 'account_id';
    const movements = await ctx.fintoc.v2.accounts.movements.list({
        account_id: accountId,
    });
    let count = 0;
    for await (const movement of movements) {
        count += 1;
        t.is(movement.method, 'get');
        t.is(movement.url, `v2/accounts/${accountId}/movements`);
        t.is(movement.params.account_id, accountId);
    }
    t.true(count > 0);
});
(0, ava_1.default)('fintoc.v2.accounts.movements.get()', async (t) => {
    const ctx = t.context;
    const accountId = 'account_id';
    const movementId = 'movement_id';
    const movement = await ctx.fintoc.v2.accounts.movements.get(movementId, {
        account_id: accountId,
    });
    t.is(movement.method, 'get');
    t.is(movement.url, `v2/accounts/${accountId}/movements/${movementId}`);
    t.is(movement.params.account_id, accountId);
});
(0, ava_1.default)('fintoc.v2.accounts.update()', async (t) => {
    const ctx = t.context;
    const entityId = 'entity_123';
    const updateData = {
        description: 'Updated account description',
    };
    const account = await ctx.fintoc.v2.accounts.update(entityId, updateData);
    t.is(account.method, 'patch');
    t.is(account.url, `v2/accounts/${entityId}`);
    t.is(account.json.description, updateData.description);
});
