// fintoc/build/main/spec/jws.spec.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const crypto = __importStar(require("crypto"));
const fs = __importStar(require("fs"));
const os = __importStar(require("os"));
const path = __importStar(require("path"));
const ava_1 = __importDefault(require("ava"));
const jws_1 = require("../lib/jws");
function generateTestKeyPair() {
    return crypto.generateKeyPairSync('rsa', {
        modulusLength: 2048,
        publicKeyEncoding: { type: 'spki', format: 'pem' },
        privateKeyEncoding: { type: 'pkcs8', format: 'pem' },
    });
}
function createTempKeyFile(privateKey) {
    const tempDir = os.tmpdir();
    const tempFile = path.join(tempDir, `test-key-${Date.now()}.pem`);
    fs.writeFileSync(tempFile, privateKey);
    return tempFile;
}
function parseJWSHeader(jwsHeader) {
    const [protectedHeaderBase64] = jwsHeader.split('.');
    const protectedHeader = JSON.parse(Buffer.from(protectedHeaderBase64, 'base64url').toString());
    return protectedHeader;
}
function verifyJWSSignature(jwsHeader, payload, publicKey) {
    const [protectedHeaderBase64, signatureBase64] = jwsHeader.split('.');
    const payloadBase64 = Buffer.from(payload).toString('base64url');
    const signingInput = `${protectedHeaderBase64}.${payloadBase64}`;
    const signature = Buffer.from(signatureBase64, 'base64url');
    const verifier = crypto.createVerify('sha256');
    verifier.update(signingInput);
    return verifier.verify(publicKey, signature);
}
ava_1.default.beforeEach((t) => {
    const ctx = t.context;
    const keyPair = generateTestKeyPair();
    ctx.privateKey = keyPair.privateKey;
    ctx.publicKey = keyPair.publicKey;
    ctx.testPayload = '{"test": "data"}';
});
ava_1.default.afterEach((t) => {
    const ctx = t.context;
    if (ctx.tempKeyFile && fs.existsSync(ctx.tempKeyFile)) {
        fs.unlinkSync(ctx.tempKeyFile);
    }
});
(0, ava_1.default)('JWSSignature initialization with PEM string', (t) => {
    const ctx = t.context;
    const jws = new jws_1.JWSSignature(ctx.privateKey);
    t.truthy(jws);
});
(0, ava_1.default)('JWSSignature initialization with PEM buffer', (t) => {
    const ctx = t.context;
    const keyBuffer = Buffer.from(ctx.privateKey, 'utf8');
    const jws = new jws_1.JWSSignature(keyBuffer);
    t.truthy(jws);
});
(0, ava_1.default)('JWSSignature initialization with file path', (t) => {
    const ctx = t.context;
    const tempKeyFile = createTempKeyFile(ctx.privateKey);
    ctx.tempKeyFile = tempKeyFile;
    const jws = new jws_1.JWSSignature(tempKeyFile);
    t.truthy(jws);
});
(0, ava_1.default)('JWSSignature throws error with invalid input type', (t) => {
    t.throws(() => {
        // eslint-disable-next-line no-new
        new jws_1.JWSSignature(123);
    }, { message: /Private key must be a string/ });
});
(0, ava_1.default)('JWSSignature throws error with non-existent file path', (t) => {
    t.throws(() => {
        // eslint-disable-next-line no-new
        new jws_1.JWSSignature('/non/existent/path.pem');
    }, { message: /Failed to read private key file/ });
});
(0, ava_1.default)('JWSSignature throws error with invalid key format', (t) => {
    t.throws(() => {
        // eslint-disable-next-line no-new
        new jws_1.JWSSignature('invalid-key-content');
    }, { message: /Invalid private key format/ });
});
(0, ava_1.default)('generateHeader creates valid JWS header structure', (t) => {
    const ctx = t.context;
    const jws = new jws_1.JWSSignature(ctx.privateKey);
    const header = jws.generateHeader(ctx.testPayload);
    t.is(header.split('.').length, 2);
    const protectedHeader = parseJWSHeader(header);
    t.is(protectedHeader.alg, 'RS256');
    t.truthy(protectedHeader.nonce);
    t.truthy(protectedHeader.ts);
    t.deepEqual(protectedHeader.crit, ['ts', 'nonce']);
});
(0, ava_1.default)('generateHeader creates valid cryptographic signature', (t) => {
    const ctx = t.context;
    const jws = new jws_1.JWSSignature(ctx.privateKey);
    const header = jws.generateHeader(ctx.testPayload);
    const isValid = verifyJWSSignature(header, ctx.testPayload, ctx.publicKey);
    t.true(isValid);
});
(0, ava_1.default)('generateHeader timestamp is current', (t) => {
    const ctx = t.context;
    const jws = new jws_1.JWSSignature(ctx.privateKey);
    const header = jws.generateHeader(ctx.testPayload);
    const protectedHeader = parseJWSHeader(header);
    const headerTimestamp = protectedHeader.ts;
    const currentTimestamp = Math.floor(Date.now() / 1000);
    t.true(Math.abs(currentTimestamp - headerTimestamp) <= 10);
});
(0, ava_1.default)('generateHeader creates unique nonces', (t) => {
    const ctx = t.context;
    const jws = new jws_1.JWSSignature(ctx.privateKey);
    const header1 = jws.generateHeader(ctx.testPayload);
    const header2 = jws.generateHeader(ctx.testPayload);
    const protectedHeader1 = parseJWSHeader(header1);
    const protectedHeader2 = parseJWSHeader(header2);
    t.not(protectedHeader1.nonce, protectedHeader2.nonce);
});
(0, ava_1.default)('generateHeader works with empty payload', (t) => {
    const ctx = t.context;
    const jws = new jws_1.JWSSignature(ctx.privateKey);
    const header = jws.generateHeader('');
    const isValid = verifyJWSSignature(header, '', ctx.publicKey);
    t.true(isValid);
});
(0, ava_1.default)('generateHeader works with complex JSON payload', (t) => {
    const ctx = t.context;
    const complexPayload = JSON.stringify({
        user: 'test@example.com',
        data: { nested: { value: 123 }, array: [1, 2, 3] },
        timestamp: Date.now(),
    });
    const jws = new jws_1.JWSSignature(ctx.privateKey);
    const header = jws.generateHeader(complexPayload);
    const isValid = verifyJWSSignature(header, complexPayload, ctx.publicKey);
    t.true(isValid);
});
(0, ava_1.default)('nonce is hexadecimal string of correct length', (t) => {
    const ctx = t.context;
    const jws = new jws_1.JWSSignature(ctx.privateKey);
    const header = jws.generateHeader(ctx.testPayload);
    const protectedHeader = parseJWSHeader(header);
    const { nonce } = protectedHeader;
    t.is(nonce.length, 32);
    t.regex(nonce, /^[0-9a-f]+$/);
});
