// fintoc/build/main/spec/mixins/managerMixin/mocks/complexMockManager.js
"use strict";
/* eslint-disable no-console, class-methods-use-this, @typescript-eslint/no-unused-vars */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComplexMockManager = void 0;
const mixins_1 = require("../../../../lib/mixins");
class ComplexMockManager extends mixins_1.ManagerMixin {
    postListHandler(objects, args) {
        console.log('Executing the "post list" handler');
        return objects;
    }
    postGetHandler(object, identifier, args) {
        console.log('Executing the "post get" handler');
        return object;
    }
    postCreateHandler(object, args) {
        console.log('Executing the "post create" handler');
        return object;
    }
    postUpdateHandler(object, identifier, args) {
        console.log('Executing the "post update" handler');
        return object;
    }
    postDeleteHandler(identifier, args) {
        console.log('Executing the "post delete" handler');
        return identifier;
    }
}
exports.ComplexMockManager = ComplexMockManager;
ComplexMockManager.resource = 'this_resource_does_not_exist';
ComplexMockManager.methods = ['list', 'get', 'create', 'update', 'delete'];
