// fintoc/build/main/lib/resources/link.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Link_accountsManager, _Link_subscriptionsManager, _Link_taxReturnsManager, _Link_invoicesManager, _Link_refreshIntentsManager, _Link_linkToken;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Link = void 0;
const managers_1 = require("../managers");
const resourceMixin_1 = require("../mixins/resourceMixin");
class Link extends resourceMixin_1.ResourceMixin {
    constructor(client, handlers, methods, path, content) {
        super(client, handlers, methods, path, content);
        _Link_accountsManager.set(this, void 0);
        _Link_subscriptionsManager.set(this, void 0);
        _Link_taxReturnsManager.set(this, void 0);
        _Link_invoicesManager.set(this, void 0);
        _Link_refreshIntentsManager.set(this, void 0);
        // @ts-ignore: declared but value is never used
        _Link_linkToken.set(this, void 0);
        __classPrivateFieldSet(this, _Link_accountsManager, undefined, "f");
        __classPrivateFieldSet(this, _Link_subscriptionsManager, undefined, "f");
        __classPrivateFieldSet(this, _Link_taxReturnsManager, undefined, "f");
        __classPrivateFieldSet(this, _Link_invoicesManager, undefined, "f");
        __classPrivateFieldSet(this, _Link_refreshIntentsManager, undefined, "f");
    }
    _setLinkToken(newLinkToken) {
        __classPrivateFieldSet(this, _Link_linkToken, newLinkToken, "f");
    }
    get _linkToken() {
        return __classPrivateFieldGet(this, _Link_linkToken, "f");
    }
    /**
     * @deprecated This method is deprecated and will be removed in a future version.
     * Please use the Fintoc client directly to access accounts. See: https://github.com/fintoc-com/fintoc-node/pull/34
     */
    get accounts() {
        console.warn('Warning: Link.accounts is deprecated and will be removed in a future version. Please use the Fintoc client directly to access accounts. See: https://github.com/fintoc-com/fintoc-node/pull/34');
        if (__classPrivateFieldGet(this, _Link_accountsManager, "f") === undefined) {
            __classPrivateFieldSet(this, _Link_accountsManager, new managers_1.AccountsManager('/v1/accounts', this._useClient()), "f");
        }
        return __classPrivateFieldGet(this, _Link_accountsManager, "f");
    }
    set accounts(newValue) {
        throw new ReferenceError('Attribute name corresponds to a manager');
    }
    /**
     * @deprecated This method is deprecated and will be removed in a future version.
     * Please use the Fintoc client directly to access subscriptions. See: https://github.com/fintoc-com/fintoc-node/pull/34
     */
    get subscriptions() {
        console.warn('Warning: Link.subscriptions is deprecated and will be removed in a future version. Please use the Fintoc client directly to access subscriptions. See: https://github.com/fintoc-com/fintoc-node/pull/34');
        if (__classPrivateFieldGet(this, _Link_subscriptionsManager, "f") === undefined) {
            __classPrivateFieldSet(this, _Link_subscriptionsManager, new managers_1.SubscriptionsManager('/v1/subscriptions', this._useClient()), "f");
        }
        return __classPrivateFieldGet(this, _Link_subscriptionsManager, "f");
    }
    set subscriptions(newValue) {
        throw new ReferenceError('Attribute name corresponds to a manager');
    }
    /**
     * @deprecated This method is deprecated and will be removed in a future version.
     * Please use the Fintoc client directly to access taxReturns. See: https://github.com/fintoc-com/fintoc-node/pull/34
     */
    get taxReturns() {
        console.warn('Warning: Link.taxReturns is deprecated and will be removed in a future version. Please use the Fintoc client directly to access taxReturns. See: https://github.com/fintoc-com/fintoc-node/pull/34');
        if (__classPrivateFieldGet(this, _Link_taxReturnsManager, "f") === undefined) {
            __classPrivateFieldSet(this, _Link_taxReturnsManager, new managers_1.TaxReturnsManager('/v1/tax_returns', this._useClient()), "f");
        }
        return __classPrivateFieldGet(this, _Link_taxReturnsManager, "f");
    }
    set taxReturns(newValue) {
        throw new ReferenceError('Attribute name corresponds to a manager');
    }
    /**
     * @deprecated This method is deprecated and will be removed in a future version.
     * Please use the Fintoc client directly to access invoices. See: https://github.com/fintoc-com/fintoc-node/pull/34
     */
    get invoices() {
        console.warn('Warning: Link.invoices is deprecated and will be removed in a future version. Please use the Fintoc client directly to access invoices. See: https://github.com/fintoc-com/fintoc-node/pull/34');
        if (__classPrivateFieldGet(this, _Link_invoicesManager, "f") === undefined) {
            __classPrivateFieldSet(this, _Link_invoicesManager, new managers_1.InvoicesManager('/v1/invoices', this._useClient()), "f");
        }
        return __classPrivateFieldGet(this, _Link_invoicesManager, "f");
    }
    set invoices(newValue) {
        throw new ReferenceError('Attribute name corresponds to a manager');
    }
    /**
     * @deprecated This method is deprecated and will be removed in a future version.
     * Please use the Fintoc client directly to access refreshIntents. See: https://github.com/fintoc-com/fintoc-node/pull/34
     */
    get refreshIntents() {
        console.warn('Warning: Link.refreshIntents is deprecated and will be removed in a future version. Please use the Fintoc client directly to access refreshIntents. See: https://github.com/fintoc-com/fintoc-node/pull/34');
        if (__classPrivateFieldGet(this, _Link_refreshIntentsManager, "f") === undefined) {
            __classPrivateFieldSet(this, _Link_refreshIntentsManager, new managers_1.RefreshIntentsManager('/v1/refresh_intents', this._useClient()), "f");
        }
        return __classPrivateFieldGet(this, _Link_refreshIntentsManager, "f");
    }
    set refreshIntents(newValue) {
        throw new ReferenceError('Attribute name corresponds to a manager');
    }
}
exports.Link = Link;
_Link_accountsManager = new WeakMap(), _Link_subscriptionsManager = new WeakMap(), _Link_taxReturnsManager = new WeakMap(), _Link_invoicesManager = new WeakMap(), _Link_refreshIntentsManager = new WeakMap(), _Link_linkToken = new WeakMap();
Link.resourceIdentifier = '_linkToken';
