// fintoc/build/module/lib/mixins/managerMixin.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { resourceAll, resourceCreate, resourceDelete, resourceGet, resourceUpdate, } from '../resourceHandlers';
import { canRaiseHTTPError, getResourceClass } from '../utils';
export class ManagerMixin {
    static resource;
    static methods;
    #path;
    #handlers;
    _client;
    constructor(path, client) {
        this.#path = path;
        this._client = client;
        this.#handlers = {
            update: this.postUpdateHandler.bind(this),
            delete: this.postDeleteHandler.bind(this),
        };
    }
    get #originatingClass() {
        return this.constructor;
    }
    list(args) {
        if (!this.#originatingClass.methods.includes('list')) {
            throw new TypeError(`${this.#originatingClass.name}.list is not a valid function of of this manager`);
        }
        return this._list(args);
    }
    all(args) {
        console.warn("Warning: The 'all()' method is deprecated. Please use 'list()' instead.");
        if (!this.#originatingClass.methods.includes('list')) {
            throw new TypeError(`${this.#originatingClass.name}.all is not a valid function of of this manager`);
        }
        return this._list(args);
    }
    /**
     * Return an instance of the resource being handled by the manager.
     *
     * @param identifier - The identifier of the resource to get
     * @param args - Object with the arguments to filter the query, using the API parameters
     * @returns An instance of the resource
     */
    get(identifier, args) {
        if (!this.#originatingClass.methods.includes('get')) {
            throw new TypeError(`${this.#originatingClass.name}.get is not a valid function of of this manager`);
        }
        return this._get(identifier, args);
    }
    /**
     * Create an instance of the resource being handled by the manager.
     *
     * @param args - Data to be passed for the object creation, as specified by the API
     * @returns The created instance of the object
     */
    create(args) {
        if (!this.#originatingClass.methods.includes('create')) {
            throw new TypeError(`${this.#originatingClass.name}.create is not a valid function of of this manager`);
        }
        return this._create(args);
    }
    /**
     * Update an instance of the resource being handled by the manager.
     *
     * @param identifier - The identifier of the resource to update
     * @param args - Data to be passed for the object to be updated with, as specified by the API
     * @returns The updated instance of the object
     */
    update(identifier, args) {
        if (!this.#originatingClass.methods.includes('update')) {
            throw new TypeError(`${this.#originatingClass.name}.update is not a valid function of of this manager`);
        }
        return this._update(identifier, args);
    }
    /**
     * Delete an instance of the resource being handled by the manager.
     *
     * @param identifier - The identifier of the resource to delete
     * @param args - Object with the arguments to filter the query, using the API parameters
     * @returns The identifier of the deleted object
     */
    delete(identifier, args) {
        if (!this.#originatingClass.methods.includes('delete')) {
            throw new TypeError(`${this.#originatingClass.name}.delete is not a valid function of of this manager`);
        }
        return this._delete(identifier, args);
    }
    async _list(args) {
        const innerArgs = args || {};
        const klass = await getResourceClass(this.#originatingClass.resource);
        const objects = await resourceAll(this._client, this.buildPath(innerArgs), klass, this.#handlers, this.#originatingClass.methods, innerArgs);
        return this.postListHandler(objects, innerArgs);
    }
    async _get(identifier, args) {
        const innerArgs = args || {};
        const klass = await getResourceClass(this.#originatingClass.resource);
        const object = await resourceGet(this._client, this.buildPath(innerArgs), identifier, klass, this.#handlers, this.#originatingClass.methods, innerArgs);
        return this.postGetHandler(object, identifier, innerArgs);
    }
    async _create(args) {
        const innerArgs = args || {};
        const { path_ = this.buildPath(), idempotency_key: idempotencyKey } = innerArgs;
        const klass = await getResourceClass(this.#originatingClass.resource);
        const object = await resourceCreate(this._client, path_.toString(), klass, this.#handlers, this.#originatingClass.methods, innerArgs, idempotencyKey?.toString());
        return this.postCreateHandler(object, innerArgs);
    }
    async _update(identifier, args) {
        const innerArgs = args || {};
        const { path_ } = innerArgs;
        const klass = await getResourceClass(this.#originatingClass.resource);
        const object = await resourceUpdate(this._client, this.buildPath(innerArgs), identifier, klass, this.#handlers, this.#originatingClass.methods, innerArgs, path_?.toString());
        return this.postUpdateHandler(object, identifier, innerArgs);
    }
    async _delete(identifier, args) {
        const innerArgs = args || {};
        await resourceDelete(this._client, this.buildPath(innerArgs), identifier, this.#handlers);
        return this.postDeleteHandler(identifier, innerArgs);
    }
    /* eslint-disable class-methods-use-this, @typescript-eslint/no-unused-vars */
    postListHandler(objects, args) {
        return objects;
    }
    postGetHandler(object, identifier, args) {
        return object;
    }
    postCreateHandler(object, args) {
        return object;
    }
    postUpdateHandler(object, identifier, args) {
        return object;
    }
    postDeleteHandler(identifier, args) {
        return identifier;
    }
    buildPath(args) {
        let path = this.#path;
        for (const [key, value] of Object.entries(args || {})) {
            path = path.replace(`{${key}}`, value.toString());
        }
        return path;
    }
}
__decorate([
    canRaiseHTTPError
], ManagerMixin.prototype, "_list", null);
__decorate([
    canRaiseHTTPError
], ManagerMixin.prototype, "_get", null);
__decorate([
    canRaiseHTTPError
], ManagerMixin.prototype, "_create", null);
__decorate([
    canRaiseHTTPError
], ManagerMixin.prototype, "_update", null);
__decorate([
    canRaiseHTTPError
], ManagerMixin.prototype, "_delete", null);
