// fintoc/build/main/spec/resources/account.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ava_1 = __importDefault(require("ava"));
const client_1 = require("../../lib/client");
const mixins_1 = require("../../lib/mixins");
const account_1 = require("../../lib/resources/account");
ava_1.default.beforeEach((t) => {
    const ctx = t.context;
    ctx.baseUrl = 'https://test.com';
    ctx.apiKey = 'super_secret_api_key';
    ctx.userAgent = 'fintoc-node/test';
    ctx.params = { first_param: 'first_value', second_param: 'second_value' };
    ctx.client = new client_1.Client({
        baseUrl: ctx.baseUrl,
        apiKey: ctx.apiKey,
        userAgent: ctx.userAgent,
        params: ctx.params,
    });
    ctx.path = '/accounts';
    ctx.handlers = {
        update: (object) => object,
        delete: (identifier) => identifier,
    };
});
(0, ava_1.default)('"Account" resource movements manager', async (t) => {
    const ctx = t.context;
    // @ts-ignore: property is protected
    const account = await account_1.Account._build(ctx.client, ctx.handlers, [], ctx.path, { id: 'idx' });
    t.assert(account.movements instanceof mixins_1.ManagerMixin);
    await t.throwsAsync(async () => {
        account.movements = null;
    }, { instanceOf: ReferenceError });
    t.not(account.movements, null);
    t.assert(account.movements instanceof mixins_1.ManagerMixin);
});
