// fintoc/build/main/spec/resources/link.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ava_1 = __importDefault(require("ava"));
const client_1 = require("../../lib/client");
const mixins_1 = require("../../lib/mixins");
const link_1 = require("../../lib/resources/link");
ava_1.default.beforeEach((t) => {
    const ctx = t.context;
    ctx.baseUrl = 'https://test.com';
    ctx.apiKey = 'super_secret_api_key';
    ctx.userAgent = 'fintoc-node/test';
    ctx.params = { first_param: 'first_value', second_param: 'second_value' };
    ctx.client = new client_1.Client({
        baseUrl: ctx.baseUrl,
        apiKey: ctx.apiKey,
        userAgent: ctx.userAgent,
        params: ctx.params,
    });
    ctx.path = '/links';
    ctx.handlers = {
        update: (object) => object,
        delete: (identifier) => identifier,
    };
});
(0, ava_1.default)('"Link" resource accounts manager', async (t) => {
    const ctx = t.context;
    // @ts-ignore: property is protected
    const link = await link_1.Link._build(ctx.client, ctx.handlers, [], ctx.path, {});
    t.assert(link.accounts instanceof mixins_1.ManagerMixin);
    await t.throwsAsync(async () => {
        link.accounts = null;
    }, { instanceOf: ReferenceError });
    t.not(link.accounts, null);
    t.assert(link.accounts instanceof mixins_1.ManagerMixin);
});
(0, ava_1.default)('"Link" resource subscriptions manager', async (t) => {
    const ctx = t.context;
    // @ts-ignore: property is protected
    const link = await link_1.Link._build(ctx.client, ctx.handlers, [], ctx.path, {});
    t.assert(link.subscriptions instanceof mixins_1.ManagerMixin);
    await t.throwsAsync(async () => {
        link.subscriptions = null;
    }, { instanceOf: ReferenceError });
    t.not(link.subscriptions, null);
    t.assert(link.subscriptions instanceof mixins_1.ManagerMixin);
});
(0, ava_1.default)('"Link" resource tax returns manager', async (t) => {
    const ctx = t.context;
    // @ts-ignore: property is protected
    const link = await link_1.Link._build(ctx.client, ctx.handlers, [], ctx.path, {});
    t.assert(link.taxReturns instanceof mixins_1.ManagerMixin);
    await t.throwsAsync(async () => {
        link.taxReturns = null;
    }, { instanceOf: ReferenceError });
    t.not(link.taxReturns, null);
    t.assert(link.taxReturns instanceof mixins_1.ManagerMixin);
});
(0, ava_1.default)('"Link" resource invoices manager', async (t) => {
    const ctx = t.context;
    // @ts-ignore: property is protected
    const link = await link_1.Link._build(ctx.client, ctx.handlers, [], ctx.path, {});
    t.assert(link.invoices instanceof mixins_1.ManagerMixin);
    await t.throwsAsync(async () => {
        link.invoices = null;
    }, { instanceOf: ReferenceError });
    t.not(link.invoices, null);
    t.assert(link.invoices instanceof mixins_1.ManagerMixin);
});
(0, ava_1.default)('"Link" resource refresh intents manager', async (t) => {
    const ctx = t.context;
    // @ts-ignore: property is protected
    const link = await link_1.Link._build(ctx.client, ctx.handlers, [], ctx.path, {});
    t.assert(link.refreshIntents instanceof mixins_1.ManagerMixin);
    await t.throwsAsync(async () => {
        link.refreshIntents = null;
    }, { instanceOf: ReferenceError });
    t.not(link.refreshIntents, null);
    t.assert(link.refreshIntents instanceof mixins_1.ManagerMixin);
});
