// fintoc/build/main/spec/mocks/client/axiosInstance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockAxiosInstance = void 0;
exports.createAxiosInstance = createAxiosInstance;
const axios_1 = require("axios");
const response_1 = require("./response");
class MockAxiosInstance extends axios_1.Axios {
    constructor(config) {
        super(config);
        const { method, url, params, data, } = config || {};
        this.method = method;
        this.baseURL = url;
        this.params = params;
        this.data = data;
    }
    // @ts-ignore: return type is not assignable
    async request(config) {
        const { method, url, params, data, headers, } = config;
        const usingMethod = method || 'get';
        const usingBaseURL = this.baseURL || '';
        const query = (url === null || url === void 0 ? void 0 : url.includes('?')) ? url.split('?').slice(-1)[0].split('&') : [];
        const innerParams = Object.fromEntries(query.map((x) => x.split('=')));
        const completeParams = { ...innerParams, ...(params === undefined ? {} : params) };
        const usableURL = (url === null || url === void 0 ? void 0 : url.split('//').slice(-1)[0].split('/').slice(1).join('/').split('?')[0]) || '';
        return new response_1.MockResponse({
            method: usingMethod,
            baseURL: usingBaseURL,
            url: usableURL,
            params: completeParams,
            json: JSON.parse(data || '{}'),
            headers: headers || {},
        });
    }
}
exports.MockAxiosInstance = MockAxiosInstance;
function createAxiosInstance(config) {
    return new MockAxiosInstance(config);
}
