// fintoc/build/main/lib/client.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _a, _Client_client;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Client = void 0;
const crypto_1 = __importDefault(require("crypto"));
const axios_1 = __importDefault(require("axios"));
const jws_1 = require("./jws");
const paginator_1 = require("./paginator");
class Client {
    constructor(options) {
        const { baseUrl, apiKey, userAgent, params = {}, jwsPrivateKey, } = options;
        this.baseUrl = baseUrl;
        this.apiKey = apiKey;
        this.userAgent = userAgent;
        this.params = params || {};
        if (jwsPrivateKey) {
            this.jws = new jws_1.JWSSignature(jwsPrivateKey);
        }
    }
    get client() {
        if (__classPrivateFieldGet(_a, _a, "f", _Client_client) === undefined) {
            __classPrivateFieldSet(_a, _a, axios_1.default.create({
                baseURL: this.baseUrl,
                headers: this.staticHeaders,
                params: this.params,
            }), "f", _Client_client);
        }
        return __classPrivateFieldGet(_a, _a, "f", _Client_client);
    }
    get staticHeaders() {
        return { Authorization: this.apiKey, 'User-Agent': this.userAgent, 'Content-Type': 'application/json' };
    }
    async request(options) {
        const { path, json, paginated = false, method = 'get', params = {}, idempotencyKey, } = options;
        const url = `${this.baseUrl}${path}`;
        const allParams = { ...this.params, ...params };
        if (paginated) {
            return (0, paginator_1.paginate)({
                client: this.client,
                path: url,
                headers: this.staticHeaders,
                params: allParams,
            });
        }
        const rawBody = JSON.stringify(json);
        const headers = this.buildHeaders(method, rawBody, idempotencyKey);
        const response = await this.client.request({
            method, url, params: allParams, data: rawBody, headers,
        });
        return response.data;
    }
    extend(extension) {
        const { baseUrl, apiKey, userAgent, params, } = (extension || {});
        return new _a({
            baseUrl: baseUrl || this.baseUrl,
            apiKey: apiKey || this.apiKey,
            userAgent: userAgent || this.userAgent,
            params: params ? { ...this.params, ...params } : { ...this.params },
        });
    }
    buildHeaders(method, rawBody, idempotencyKey) {
        const headers = {
            ...this.staticHeaders,
            ...this.buildJWSHeaders(method, rawBody),
        };
        if (method === 'post') {
            headers['Idempotency-Key'] = idempotencyKey || crypto_1.default.randomUUID();
        }
        return headers;
    }
    buildJWSHeaders(method, rawBody) {
        if (this.jws && (method === 'post' || method === 'put' || method === 'patch')) {
            const signature = this.jws.generateHeader(rawBody);
            return { 'Fintoc-JWS-Signature': signature };
        }
        return {};
    }
}
exports.Client = Client;
_a = Client;
_Client_client = { value: void 0 };
