// fintoc/build/main/lib/core.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Fintoc_client;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Fintoc = void 0;
/* eslint-disable max-classes-per-file */
const client_1 = require("./client");
const constants_1 = require("./constants");
const managers_1 = require("./managers");
const v2_1 = require("./managers/v2");
const version_1 = require("./version");
class FintocV2 {
    constructor(client) {
        this.accounts = new v2_1.AccountsManager('/v2/accounts', client);
        this.transfers = new v2_1.TransfersManager('/v2/transfers', client);
        this.accountNumbers = new v2_1.AccountNumbersManager('/v2/account_numbers', client);
        this.simulate = new v2_1.SimulateManager('/v2/simulate', client);
        this.entities = new v2_1.EntitiesManager('/v2/entities', client);
        this.accountVerifications = new v2_1.AccountVerificationsManager('/v2/account_verifications', client);
    }
}
class Fintoc {
    constructor(apiKey, jwsPrivateKey) {
        _Fintoc_client.set(this, void 0);
        __classPrivateFieldSet(this, _Fintoc_client, new client_1.Client({
            baseUrl: `${constants_1.API_BASE_URL}`,
            userAgent: `fintoc-node/${version_1.version}`,
            apiKey,
            jwsPrivateKey,
        }), "f");
        this.accounts = new managers_1.AccountsManager('/v1/accounts', __classPrivateFieldGet(this, _Fintoc_client, "f"));
        this.charges = new managers_1.ChargesManager('/v1/charges', __classPrivateFieldGet(this, _Fintoc_client, "f"));
        this.checkoutSessions = new managers_1.CheckoutSessionsManager('/v1/checkout_sessions', __classPrivateFieldGet(this, _Fintoc_client, "f"));
        this.invoices = new managers_1.InvoicesManager('/v1/invoices', __classPrivateFieldGet(this, _Fintoc_client, "f"));
        this.links = new managers_1.LinksManager('/v1/links', __classPrivateFieldGet(this, _Fintoc_client, "f"));
        this.paymentIntents = new managers_1.PaymentIntentsManager('/v1/payment_intents', __classPrivateFieldGet(this, _Fintoc_client, "f"));
        this.paymentLinks = new managers_1.PaymentLinksManager('/v1/payment_links', __classPrivateFieldGet(this, _Fintoc_client, "f"));
        this.refreshIntents = new managers_1.RefreshIntentsManager('/v1/refresh_intents', __classPrivateFieldGet(this, _Fintoc_client, "f"));
        this.refunds = new managers_1.RefundsManager('/v1/refunds', __classPrivateFieldGet(this, _Fintoc_client, "f"));
        this.subscriptions = new managers_1.SubscriptionsManager('/v1/subscriptions', __classPrivateFieldGet(this, _Fintoc_client, "f"));
        this.subscriptionIntents = new managers_1.SubscriptionIntentsManager('/v1/subscription_intents', __classPrivateFieldGet(this, _Fintoc_client, "f"));
        this.taxReturns = new managers_1.TaxReturnsManager('/v1/tax_returns', __classPrivateFieldGet(this, _Fintoc_client, "f"));
        this.webhookEndpoints = new managers_1.WebhookEndpointsManager('/v1/webhook_endpoints', __classPrivateFieldGet(this, _Fintoc_client, "f"));
        this.v2 = new FintocV2(__classPrivateFieldGet(this, _Fintoc_client, "f"));
    }
}
exports.Fintoc = Fintoc;
_Fintoc_client = new WeakMap();
