// fintoc/build/main/lib/errors.js
"use strict";
/* eslint-disable max-classes-per-file */
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebhookSignatureError = exports.InvalidRequestError = exports.InstitutionError = exports.LinkError = exports.AuthenticationError = exports.ApiError = exports.FintocError = void 0;
class FintocError extends Error {
    constructor(errorData) {
        const errorType = errorData.type;
        const errorCode = errorData.code;
        const errorMessage = errorData.message;
        const errorParam = errorData.param;
        const errorDocURL = errorData.doc_url;
        let message = errorType;
        message += errorCode ? `: ${errorCode}` : '';
        message += errorParam ? ` (${errorParam})` : '';
        message += `\n${errorMessage}`;
        message += errorDocURL ? `\nCheck the docs for more info: ${errorDocURL}` : '';
        super(message);
    }
}
exports.FintocError = FintocError;
class ApiError extends FintocError {
}
exports.ApiError = ApiError;
class AuthenticationError extends FintocError {
}
exports.AuthenticationError = AuthenticationError;
class LinkError extends FintocError {
}
exports.LinkError = LinkError;
class InstitutionError extends FintocError {
}
exports.InstitutionError = InstitutionError;
class InvalidRequestError extends FintocError {
}
exports.InvalidRequestError = InvalidRequestError;
class WebhookSignatureError extends Error {
}
exports.WebhookSignatureError = WebhookSignatureError;
