// fintoc/build/main/lib/jws.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.JWSSignature = void 0;
const crypto = __importStar(require("crypto"));
const fs = __importStar(require("fs"));
/**
 * Class to handle JWS signature generation for Fintoc API requests.
 */
class JWSSignature {
    /**
     * Create a new JWSSignature instance with a private key.
     *
     * @param privateKey - The private key in one of these formats:
     *   - String containing PEM-formatted key
     *   - Buffer containing PEM-formatted key
     *   - String path to a PEM key file
     */
    constructor(privateKey) {
        this.privateKey = JWSSignature.loadPrivateKey(privateKey);
    }
    /**
     * Generate a JWS signature header for the given payload
     * @param rawBody - The request body as a string
     * @returns The JWS signature header to be used in 'Fintoc-JWS-Signature' header.
     */
    generateHeader(rawBody) {
        const headers = {
            alg: 'RS256',
            nonce: JWSSignature.generateNonce(),
            ts: Math.floor(Date.now() / 1000),
            crit: ['ts', 'nonce'],
        };
        const protectedBase64 = Buffer.from(JSON.stringify(headers)).toString('base64url');
        const payloadBase64 = Buffer.from(rawBody).toString('base64url');
        const signingInput = `${protectedBase64}.${payloadBase64}`;
        const signature = this.createSignature(signingInput);
        const signatureBase64 = Buffer.from(signature).toString('base64url');
        return `${protectedBase64}.${signatureBase64}`;
    }
    static loadPrivateKey(keyInput) {
        let keyMaterial;
        if (Buffer.isBuffer(keyInput)) {
            keyMaterial = keyInput;
        }
        else if (typeof keyInput === 'string') {
            if (JWSSignature.isFilePath(keyInput)) {
                try {
                    keyMaterial = fs.readFileSync(keyInput, 'utf8');
                }
                catch (error) {
                    throw new Error(`Failed to read private key file: ${keyInput}. ${error}`);
                }
            }
            else {
                keyMaterial = keyInput;
            }
        }
        else {
            throw new Error('Private key must be a string (PEM content or file path) or Buffer');
        }
        try {
            return crypto.createPrivateKey(keyMaterial);
        }
        catch (error) {
            throw new Error(`Invalid private key format: ${error}`);
        }
    }
    static isFilePath(input) {
        if (!input || typeof input !== 'string') {
            return false;
        }
        const trimmedInput = input.trim();
        if (trimmedInput.includes('-----BEGIN') && trimmedInput.includes('-----END')) {
            return false;
        }
        if (trimmedInput.includes('\n') || trimmedInput.includes('\r')) {
            return false;
        }
        if (trimmedInput.length > 1000) {
            return false;
        }
        const hasPathSeparators = trimmedInput.includes('/') || trimmedInput.includes('\\');
        const hasFileExtension = /\.[a-zA-Z0-9]{1,10}$/.test(trimmedInput);
        const looksLikeAbsolutePath = /^\/|^[a-zA-Z]:[\\\\//]|^~/.test(trimmedInput);
        return hasPathSeparators || (hasFileExtension && looksLikeAbsolutePath);
    }
    static generateNonce() {
        return crypto.randomBytes(16).toString('hex');
    }
    createSignature(data) {
        return crypto.createSign('sha256')
            .update(data)
            .sign({ key: this.privateKey, padding: crypto.constants.RSA_PKCS1_PADDING });
    }
}
exports.JWSSignature = JWSSignature;
