// fintoc/build/main/lib/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toTitle = toTitle;
exports.snakeToPascal = snakeToPascal;
exports.singularize = singularize;
exports.isISODate = isISODate;
exports.getResourcesModule = getResourcesModule;
exports.getErrorsModule = getErrorsModule;
exports.getResourceClass = getResourceClass;
exports.getErrorClass = getErrorClass;
exports.canRaiseHTTPError = canRaiseHTTPError;
exports.serialize = serialize;
exports.objetize = objetize;
exports.objetizeGenerator = objetizeGenerator;
const axios_1 = __importDefault(require("axios"));
const errors = __importStar(require("./errors"));
const genericFintocResource_1 = require("./resources/genericFintocResource");
/**
 * Transform a string into title case.
 *
 * @param rawString - String to be transformed into a title
 * @returns Transformed title string
 */
function toTitle(rawString) {
    return rawString[0].toUpperCase() + rawString.substr(1).toLowerCase();
}
/**
 * Transforms a snake-cased string into a pascal-cased one.
 *
 * @param snakeString - Snake-cased string to be transformed
 * @returns Pascal-cased string
 */
function snakeToPascal(snakeString) {
    return snakeString.split('_').map(toTitle).join('');
}
/**
 * Remove the last 's' from a string if exists.
 *
 * @param rawString - String to be singularized
 * @returns The singularized string
 */
function singularize(rawString) {
    return rawString.replace(/s$/, '');
}
/**
 * Tries to parse an ISO formatted string to check if it is parseable as a Date object.
 *
 * @param rawDate - ISO formatted string to be checked
 * @returns A boolean that represents if `rawDate` is parseable as a Date object
 */
function isISODate(rawDate) {
    return !Number.isNaN(Date.parse(rawDate));
}
/**
 * Gets and returns the resources module interfaced as an IModule.
 *
 * @param resourceName - Pascal cased string with the name of the resource
 * @returns The resources module interfaced as an IModule
 */
async function getResourcesModule(resourceName) {
    const camelCaseName = resourceName.charAt(0).toLowerCase() + resourceName.slice(1);
    try {
        const resources = await Promise.resolve(`${`./resources/${camelCaseName}`}`).then(s => __importStar(require(s)));
        return resources;
    }
    catch (_a) {
        return {};
    }
}
/**
 * Gets and returns the errors module interfaced as an IModule.
 *
 * @returns The errors module interfaced as an IModule
 */
function getErrorsModule() {
    return errors;
}
/**
 * Get the class corresponding to the the resource name and value.
 *
 * @param snakeResourceName - Name of the resource in snake case
 * @param value - Value of the resource from which to get the corresponding class
 * @returns The class corresponding to the resource
 */
async function getResourceClass(snakeResourceName, value = {}) {
    const klass = value === null || value === void 0 ? void 0 : value.constructor;
    if (klass === Object) {
        const resourceName = snakeToPascal(snakeResourceName);
        const resourcesModule = await getResourcesModule(resourceName);
        return resourcesModule[resourceName] || genericFintocResource_1.GenericFintocResource;
    }
    if ((klass === String) && isISODate(value)) {
        return Date;
    }
    return klass;
}
/**
 * Get the error corresponding to the error name.
 *
 * @param snakeErrorName - Name of the error in snake case
 * @returns The class corresponding to the error
 */
function getErrorClass(snakeErrorName) {
    const errorsModule = getErrorsModule();
    const errorName = snakeToPascal(snakeErrorName);
    return errorsModule[errorName] || errorsModule.FintocError;
}
/**
 * Decorator designed for an instance method. It tries tp execute the function,
 * catches exceptions and re-rises the adequate Fintoc exception.
 *
 * @example
 * ```ts
 * class SampleClass {
 *   @canRaiseHTTPError
 *   myFunction() {
 *     console.log('This method is decorated by the function.');
 *   }
 * }
 * ```
 */
function canRaiseHTTPError(_, __, descriptor) {
    const newDescriptor = { ...descriptor };
    newDescriptor.value = async function wrapper(...args) {
        try {
            return await descriptor.value.apply(this, args);
        }
        catch (exc) {
            if (axios_1.default.isAxiosError(exc) && exc.response) {
                const errorData = exc.response.data;
                const ErrorKlass = getErrorClass(errorData.error.type);
                throw new ErrorKlass(errorData.error);
            }
            else {
                throw exc;
            }
        }
    };
    return newDescriptor;
}
/**
 * Serializes an object.
 *
 * @param object - Object to be serialized
 * @returns The serialized object (object with only JSON-serializable fields)
 */
function serialize(object) {
    if (typeof (object === null || object === void 0 ? void 0 : object.serialize) === 'function') {
        return object.serialize();
    }
    if ((object === null || object === void 0 ? void 0 : object.constructor) === Date) {
        return object.toISOString();
    }
    return object;
}
/**
 * Wraps some data with a class.
 *
 * @param Klass - Class that will wrap the data
 * @param client - The Client object passed to the newly created object
 * @param data - The data that the new object will contain
 * @param handlers - The post-request handlers
 * @param methods - An array of the methods that the object can execute
 * @param path - The path within the API server for the resource
 * @returns The data wrapped on the corresponding class
 */
async function objetize(Klass, client, data, handlers = {}, methods = [], path = null) {
    if (data === null) {
        return null;
    }
    if ([String, Number, Object, Boolean].includes(Klass)) {
        return Klass(data);
    }
    if (Klass === Date) {
        return new Klass(data);
    }
    return Klass._build(client, handlers, methods, path, data);
}
/**
 * Wraps and yields some data (from a generator) with a class.
 *
 * @param generator - The generator that yields elements to be wrapped
 * @param klass - Class that will wrap the data
 * @param client - The Client object passed to the newly created objects
 * @param handlers - The post-request handlers
 * @param methods - An array of the methods that the objects can execute
 * @param path - The path within the API server for the resource
 * @returns A generator producing a sequence of data wrapped with the class
 */
async function* objetizeGenerator(generator, klass, client, handlers = {}, methods = [], path = null) {
    for await (const element of generator) {
        yield objetize(klass, client, element, handlers, methods, path);
    }
}
