// fintoc/build/main/lib/webhook.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebhookSignature = void 0;
const crypto = __importStar(require("crypto"));
const errors_1 = require("./errors");
const DEFAULT_TOLERANCE_SECONDS = 300;
class WebhookSignature {
    /**
     * Verifies the authenticity of incoming webhooks from Fintoc by validating the signature.
     *
     * @param payload - The raw request body as a string or Buffer
     * @param header - The Fintoc-Signature header value
     * @param secret - Your webhook secret key (found in your Fintoc dashboard)
     * @param tolerance - Number of seconds to tolerate when checking timestamp (default: 300)
     * @throws {WebhookSignatureError} If the signature is invalid or timestamp is outside
     * tolerance window
     */
    static verifyHeader(payload, header, secret, tolerance = DEFAULT_TOLERANCE_SECONDS) {
        if (!header) {
            throw new errors_1.WebhookSignatureError('Missing Fintoc-Signature header.');
        }
        if (!secret) {
            throw new errors_1.WebhookSignatureError('Webhook secret is required for verification.');
        }
        const { timestamp, signature } = WebhookSignature.parseHeader(header);
        WebhookSignature.verifyTimestamp(timestamp, tolerance);
        const expectedSignature = WebhookSignature.computeSignature(payload, timestamp, secret);
        if (!WebhookSignature.secureCompare(signature, expectedSignature)) {
            throw new errors_1.WebhookSignatureError('Signature mismatch. The webhook signature is not valid.');
        }
    }
    static parseHeader(header) {
        const parts = header.split(',');
        let timestampStr;
        let signature;
        for (const part of parts) {
            const [key, value] = part.split('=');
            if (key === 't') {
                timestampStr = value;
            }
            else if (key === 'v1') {
                signature = value;
            }
        }
        if (!timestampStr || !signature) {
            throw new errors_1.WebhookSignatureError('Unable to extract timestamp and signatures from header');
        }
        const timestamp = parseInt(timestampStr, 10);
        if (Number.isNaN(timestamp)) {
            throw new errors_1.WebhookSignatureError('Unable to extract timestamp and signatures from header');
        }
        return { timestamp, signature };
    }
    static verifyTimestamp(timestamp, toleranceSeconds) {
        const nowSeconds = Math.floor(Date.now() / 1000);
        const timeDifference = Math.abs(nowSeconds - timestamp);
        if (timeDifference > toleranceSeconds) {
            throw new errors_1.WebhookSignatureError(`Timestamp outside the tolerance zone (${timestamp})`);
        }
    }
    static computeSignature(payload, timestamp, secret) {
        const signedPayload = `${timestamp}.${payload.toString('utf8')}`;
        return crypto
            .createHmac('sha256', secret)
            .update(signedPayload)
            .digest('hex');
    }
    static secureCompare(a, b) {
        if (a.length !== b.length) {
            return false;
        }
        return crypto.timingSafeEqual(Buffer.from(a), Buffer.from(b));
    }
}
exports.WebhookSignature = WebhookSignature;
