// fintoc/build/main/spec/utils/canRaiseHTTPError.spec.js
"use strict";
/* eslint-disable class-methods-use-this */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ava_1 = __importDefault(require("ava"));
const errors_1 = require("../../lib/errors");
const utils_1 = require("../../lib/utils");
class Errors {
    noError() { } /* eslint-disable-line @typescript-eslint/no-empty-function */
    throwHTTPError() {
        const error = new Error();
        // @ts-ignore: property does not exist
        error.isAxiosError = true;
        // @ts-ignore: property does not exist
        error.response = {};
        // @ts-ignore: property does not exist
        error.response.data = {
            error: {
                type: 'api_error',
                message: 'This is a test error message',
            },
        };
        throw error;
    }
    throwGenericError() {
        throw new Error();
    }
}
__decorate([
    utils_1.canRaiseHTTPError
], Errors.prototype, "noError", null);
__decorate([
    utils_1.canRaiseHTTPError
], Errors.prototype, "throwHTTPError", null);
__decorate([
    utils_1.canRaiseHTTPError
], Errors.prototype, "throwGenericError", null);
ava_1.default.beforeEach((t) => {
    const ctx = t.context;
    ctx.errors = new Errors();
});
(0, ava_1.default)('"canRaiseHTTPError" no error', async (t) => {
    const ctx = t.context;
    await t.notThrowsAsync(ctx.errors.noError);
});
(0, ava_1.default)('"canRaiseHTTPError" HTTP error', async (t) => {
    const ctx = t.context;
    await t.throwsAsync(ctx.errors.throwHTTPError, { instanceOf: errors_1.FintocError });
});
(0, ava_1.default)('"canRaiseHTTPError" generic error', async (t) => {
    const ctx = t.context;
    await t.throwsAsync(ctx.errors.throwGenericError, { instanceOf: Error });
});
