// fintoc/build/main/spec/mixins/resourceMixin/creation.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ava_1 = __importDefault(require("ava"));
const client_1 = require("../../../lib/client");
const mixins_1 = require("../../../lib/mixins");
const genericFintocResource_1 = require("../../../lib/resources/genericFintocResource");
const link_1 = require("../../../lib/resources/link");
const initializers_1 = require("../../mocks/initializers");
const complexMockResource_1 = require("./mocks/complexMockResource");
const emptyMockResource_1 = require("./mocks/emptyMockResource");
ava_1.default.before((t) => {
    const ctx = t.context;
    ctx.axiosMock = (0, initializers_1.mockAxios)();
});
ava_1.default.after((t) => {
    const ctx = t.context;
    (0, initializers_1.restore)(ctx.axiosMock);
});
ava_1.default.beforeEach((t) => {
    const ctx = t.context;
    ctx.baseUrl = 'https://test.com';
    ctx.apiKey = 'super_secret_api_key';
    ctx.userAgent = 'fintoc-node/test';
    ctx.params = { first_param: 'first_value', second_param: 'second_value' };
    ctx.client = new client_1.Client({
        baseUrl: ctx.baseUrl,
        apiKey: ctx.apiKey,
        userAgent: ctx.userAgent,
        params: ctx.params,
    });
    ctx.path = '/resources';
    ctx.handlers = {
        update: (object) => object,
        delete: (identifier) => identifier,
    };
});
(0, ava_1.default)('"ResourceMixin" creation empty mock resource', async (t) => {
    const ctx = t.context;
    const methods = [];
    const data = {
        id: 'id0',
        identifier: 'identifier0',
        resources: [
            { id: 'id1', identifier: 'identifier1' },
            { id: 'id2', identifier: 'identifier2' },
        ],
        resource: { id: 'id3', identifier: 'identifier3' },
    };
    // @ts-ignore: property is protected
    const resource = await emptyMockResource_1.EmptyMockResource._build(ctx.client, ctx.handlers, methods, ctx.path, data);
    t.assert(resource instanceof mixins_1.ResourceMixin);
    t.assert(resource.resource instanceof genericFintocResource_1.GenericFintocResource);
    t.is(resource.resource.id, data.resource.id);
    t.assert(Array.isArray(resource.resources));
    resource.resources.forEach((subResource) => {
        t.assert(subResource instanceof genericFintocResource_1.GenericFintocResource);
    });
});
(0, ava_1.default)('"ResourceMixin" creation complex mock resource', async (t) => {
    const ctx = t.context;
    const methods = [];
    const data = {
        id: 'id0',
        identifier: 'identifier0',
        resources: [
            { id: 'id1', identifier: 'identifier1' },
            { id: 'id2', identifier: 'identifier2' },
        ],
        resource: { id: 'id3', identifier: 'identifier3' },
    };
    // @ts-ignore: property is protected
    const resource = await complexMockResource_1.ComplexMockResource._build(ctx.client, ctx.handlers, methods, ctx.path, data);
    t.assert(resource instanceof mixins_1.ResourceMixin);
    t.assert(resource.resource instanceof link_1.Link);
    t.is(resource.resource.id, data.resource.id);
    t.assert(Array.isArray(resource.resources));
    resource.resources.forEach((subResource) => {
        t.assert(subResource instanceof genericFintocResource_1.GenericFintocResource);
    });
});
(0, ava_1.default)('"ResourceMixin" creation update delete methods access', async (t) => {
    const ctx = t.context;
    const data = {
        id: 'id0',
        identifier: 'identifier0',
        resources: [
            { id: 'id1', identifier: 'identifier1' },
            { id: 'id2', identifier: 'identifier2' },
        ],
        resource: { id: 'id3', identifier: 'identifier3' },
    };
    // @ts-ignore: property is protected
    const deleteAndNoUpdateresource = await emptyMockResource_1.EmptyMockResource._build(ctx.client, ctx.handlers, ['delete'], ctx.path, data);
    t.assert(deleteAndNoUpdateresource instanceof mixins_1.ResourceMixin);
    await t.throwsAsync(async () => {
        await deleteAndNoUpdateresource.update();
    }, { instanceOf: TypeError });
    await t.notThrowsAsync(async () => {
        await deleteAndNoUpdateresource.delete();
    });
    // @ts-ignore: property is protected
    const updateAndNoDeleteResource = await emptyMockResource_1.EmptyMockResource._build(ctx.client, ctx.handlers, ['update'], ctx.path, data);
    t.assert(updateAndNoDeleteResource instanceof mixins_1.ResourceMixin);
    await t.throwsAsync(async () => {
        await updateAndNoDeleteResource.delete();
    }, { instanceOf: TypeError });
    await t.notThrowsAsync(async () => {
        await updateAndNoDeleteResource.update();
    });
});
