// fintoc/build/main/spec/mixins/resourceMixin/methods.spec.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ava_1 = __importDefault(require("ava"));
const client_1 = require("../../../lib/client");
const initializers_1 = require("../../mocks/initializers");
const complexMockResource_1 = require("./mocks/complexMockResource");
const emptyMockResource_1 = require("./mocks/emptyMockResource");
ava_1.default.before((t) => {
    const ctx = t.context;
    ctx.axiosMock = (0, initializers_1.mockAxios)();
});
ava_1.default.after((t) => {
    const ctx = t.context;
    (0, initializers_1.restore)(ctx.axiosMock);
});
ava_1.default.beforeEach((t) => {
    const ctx = t.context;
    ctx.baseUrl = 'https://test.com';
    ctx.apiKey = 'super_secret_api_key';
    ctx.userAgent = 'fintoc-node/test';
    ctx.params = { first_param: 'first_value', second_param: 'second_value' };
    ctx.client = new client_1.Client({
        baseUrl: ctx.baseUrl,
        apiKey: ctx.apiKey,
        userAgent: ctx.userAgent,
        params: ctx.params,
    });
    ctx.path = '/resources';
    ctx.handlers = {
        update: (object) => object,
        delete: (identifier) => identifier,
    };
});
(0, ava_1.default)('"ResourceMixin" methods empty mock resource delete method', async (t) => {
    const ctx = t.context;
    const methods = ['delete'];
    const data = {
        id: 'id0',
        identifier: 'identifier0',
        resources: [
            { id: 'id1', identifier: 'identifier1' },
            { id: 'id2', identifier: 'identifier2' },
        ],
        resource: { id: 'id3', identifier: 'identifier3' },
    };
    // @ts-ignore: property is protected
    const resource = await emptyMockResource_1.EmptyMockResource._build(ctx.client, ctx.handlers, methods, ctx.path, data);
    const identifier = await resource.delete();
    t.not(identifier, data.identifier);
    t.is(identifier, data.id);
});
(0, ava_1.default)('"ResourceMixin" methods complex mock resource delete method', async (t) => {
    const ctx = t.context;
    const methods = ['delete'];
    const data = {
        id: 'id0',
        identifier: 'identifier0',
        resources: [
            { id: 'id1', identifier: 'identifier1' },
            { id: 'id2', identifier: 'identifier2' },
        ],
        resource: { id: 'id3', identifier: 'identifier3' },
    };
    // @ts-ignore: property is protected
    const resource = await complexMockResource_1.ComplexMockResource._build(ctx.client, ctx.handlers, methods, ctx.path, data);
    const identifier = await resource.delete();
    t.not(identifier, data.id);
    t.is(identifier, data.identifier);
});
(0, ava_1.default)('"ResourceMixin" methods empty mock resource update method', async (t) => {
    const ctx = t.context;
    const methods = ['update'];
    const data = {
        id: 'id0',
        identifier: 'identifier0',
        resources: [
            { id: 'id1', identifier: 'identifier1' },
            { id: 'id2', identifier: 'identifier2' },
        ],
        resource: { id: 'id3', identifier: 'identifier3' },
    };
    // @ts-ignore: property is protected
    const resource = await emptyMockResource_1.EmptyMockResource._build(ctx.client, ctx.handlers, methods, ctx.path, data);
    await resource.update();
    t.assert(!resource.url.includes(data.identifier));
    t.assert(resource.url.includes(data.id));
});
(0, ava_1.default)('"ResourceMixin" methods complex mock resource update method', async (t) => {
    const ctx = t.context;
    const methods = ['update'];
    const data = {
        id: 'id0',
        identifier: 'identifier0',
        resources: [
            { id: 'id1', identifier: 'identifier1' },
            { id: 'id2', identifier: 'identifier2' },
        ],
        resource: { id: 'id3', identifier: 'identifier3' },
    };
    // @ts-ignore: property is protected
    const resource = await complexMockResource_1.ComplexMockResource._build(ctx.client, ctx.handlers, methods, ctx.path, data);
    await resource.update();
    t.assert(!resource.url.includes(data.id));
    t.assert(resource.url.includes(data.identifier));
});
