// fintoc/build/main/spec/mixins/resourceMixin/serialization.spec.js
"use strict";
/* eslint-disable no-console */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ava_1 = __importDefault(require("ava"));
const client_1 = require("../../../lib/client");
const initializers_1 = require("../../mocks/initializers");
const emptyMockResource_1 = require("./mocks/emptyMockResource");
ava_1.default.before((t) => {
    const ctx = t.context;
    ctx.axiosMock = (0, initializers_1.mockAxios)();
});
ava_1.default.after((t) => {
    const ctx = t.context;
    (0, initializers_1.restore)(ctx.axiosMock);
});
ava_1.default.beforeEach((t) => {
    const ctx = t.context;
    ctx.baseUrl = 'https://test.com';
    ctx.apiKey = 'super_secret_api_key';
    ctx.userAgent = 'fintoc-node/test';
    ctx.params = { first_param: 'first_value', second_param: 'second_value' };
    ctx.client = new client_1.Client({
        baseUrl: ctx.baseUrl,
        apiKey: ctx.apiKey,
        userAgent: ctx.userAgent,
        params: ctx.params,
    });
    ctx.path = '/resources';
    ctx.handlers = {
        update: (object) => {
            console.log('Calling update...');
            return object;
        },
        delete: (identifier) => {
            console.log('Calling delete...');
            return identifier;
        },
    };
});
(0, ava_1.default)('"ResourceMixin" serialization method', async (t) => {
    const ctx = t.context;
    const methods = ['delete'];
    const data = {
        id: 'id0',
        identifier: 'identifier0',
        resource: { id: 'id3', identifier: 'identifier3' },
    };
    // @ts-ignore: property is protected
    const resource = await emptyMockResource_1.EmptyMockResource._build(ctx.client, ctx.handlers, methods, ctx.path, data);
    t.deepEqual(resource.serialize(), data);
});
(0, ava_1.default)('"ResourceMixin" serialization method (array)', async (t) => {
    const ctx = t.context;
    const methods = ['delete'];
    const data = {
        id: 'id0',
        identifier: 'identifier0',
        resources: [
            { id: 'id1', identifier: 'identifier1' },
            { id: 'id2', identifier: 'identifier2' },
        ],
        resource: { id: 'id3', identifier: 'identifier3' },
    };
    // @ts-ignore: property is protected
    const resource = await emptyMockResource_1.EmptyMockResource._build(ctx.client, ctx.handlers, methods, ctx.path, data);
    t.deepEqual(resource.serialize(), data);
});
