// fintoc/build/main/lib/mixins/managerMixin.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _ManagerMixin_instances, _ManagerMixin_path, _ManagerMixin_handlers, _ManagerMixin_originatingClass_get;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ManagerMixin = void 0;
const resourceHandlers_1 = require("../resourceHandlers");
const utils_1 = require("../utils");
class ManagerMixin {
    constructor(path, client) {
        _ManagerMixin_instances.add(this);
        _ManagerMixin_path.set(this, void 0);
        _ManagerMixin_handlers.set(this, void 0);
        __classPrivateFieldSet(this, _ManagerMixin_path, path, "f");
        this._client = client;
        __classPrivateFieldSet(this, _ManagerMixin_handlers, {
            update: this.postUpdateHandler.bind(this),
            delete: this.postDeleteHandler.bind(this),
        }, "f");
    }
    list(args) {
        if (!__classPrivateFieldGet(this, _ManagerMixin_instances, "a", _ManagerMixin_originatingClass_get).methods.includes('list')) {
            throw new TypeError(`${__classPrivateFieldGet(this, _ManagerMixin_instances, "a", _ManagerMixin_originatingClass_get).name}.list is not a valid function of of this manager`);
        }
        return this._list(args);
    }
    all(args) {
        console.warn("Warning: The 'all()' method is deprecated. Please use 'list()' instead.");
        if (!__classPrivateFieldGet(this, _ManagerMixin_instances, "a", _ManagerMixin_originatingClass_get).methods.includes('list')) {
            throw new TypeError(`${__classPrivateFieldGet(this, _ManagerMixin_instances, "a", _ManagerMixin_originatingClass_get).name}.all is not a valid function of of this manager`);
        }
        return this._list(args);
    }
    /**
     * Return an instance of the resource being handled by the manager.
     *
     * @param identifier - The identifier of the resource to get
     * @param args - Object with the arguments to filter the query, using the API parameters
     * @returns An instance of the resource
     */
    get(identifier, args) {
        if (!__classPrivateFieldGet(this, _ManagerMixin_instances, "a", _ManagerMixin_originatingClass_get).methods.includes('get')) {
            throw new TypeError(`${__classPrivateFieldGet(this, _ManagerMixin_instances, "a", _ManagerMixin_originatingClass_get).name}.get is not a valid function of of this manager`);
        }
        return this._get(identifier, args);
    }
    /**
     * Create an instance of the resource being handled by the manager.
     *
     * @param args - Data to be passed for the object creation, as specified by the API
     * @returns The created instance of the object
     */
    create(args) {
        if (!__classPrivateFieldGet(this, _ManagerMixin_instances, "a", _ManagerMixin_originatingClass_get).methods.includes('create')) {
            throw new TypeError(`${__classPrivateFieldGet(this, _ManagerMixin_instances, "a", _ManagerMixin_originatingClass_get).name}.create is not a valid function of of this manager`);
        }
        return this._create(args);
    }
    /**
     * Update an instance of the resource being handled by the manager.
     *
     * @param identifier - The identifier of the resource to update
     * @param args - Data to be passed for the object to be updated with, as specified by the API
     * @returns The updated instance of the object
     */
    update(identifier, args) {
        if (!__classPrivateFieldGet(this, _ManagerMixin_instances, "a", _ManagerMixin_originatingClass_get).methods.includes('update')) {
            throw new TypeError(`${__classPrivateFieldGet(this, _ManagerMixin_instances, "a", _ManagerMixin_originatingClass_get).name}.update is not a valid function of of this manager`);
        }
        return this._update(identifier, args);
    }
    /**
     * Delete an instance of the resource being handled by the manager.
     *
     * @param identifier - The identifier of the resource to delete
     * @param args - Object with the arguments to filter the query, using the API parameters
     * @returns The identifier of the deleted object
     */
    delete(identifier, args) {
        if (!__classPrivateFieldGet(this, _ManagerMixin_instances, "a", _ManagerMixin_originatingClass_get).methods.includes('delete')) {
            throw new TypeError(`${__classPrivateFieldGet(this, _ManagerMixin_instances, "a", _ManagerMixin_originatingClass_get).name}.delete is not a valid function of of this manager`);
        }
        return this._delete(identifier, args);
    }
    async _list(args) {
        const innerArgs = args || {};
        const klass = await (0, utils_1.getResourceClass)(__classPrivateFieldGet(this, _ManagerMixin_instances, "a", _ManagerMixin_originatingClass_get).resource);
        const objects = await (0, resourceHandlers_1.resourceAll)(this._client, this.buildPath(innerArgs), klass, __classPrivateFieldGet(this, _ManagerMixin_handlers, "f"), __classPrivateFieldGet(this, _ManagerMixin_instances, "a", _ManagerMixin_originatingClass_get).methods, innerArgs);
        return this.postListHandler(objects, innerArgs);
    }
    async _get(identifier, args) {
        const innerArgs = args || {};
        const klass = await (0, utils_1.getResourceClass)(__classPrivateFieldGet(this, _ManagerMixin_instances, "a", _ManagerMixin_originatingClass_get).resource);
        const object = await (0, resourceHandlers_1.resourceGet)(this._client, this.buildPath(innerArgs), identifier, klass, __classPrivateFieldGet(this, _ManagerMixin_handlers, "f"), __classPrivateFieldGet(this, _ManagerMixin_instances, "a", _ManagerMixin_originatingClass_get).methods, innerArgs);
        return this.postGetHandler(object, identifier, innerArgs);
    }
    async _create(args) {
        const innerArgs = args || {};
        const { path_ = this.buildPath(), idempotency_key: idempotencyKey } = innerArgs;
        const klass = await (0, utils_1.getResourceClass)(__classPrivateFieldGet(this, _ManagerMixin_instances, "a", _ManagerMixin_originatingClass_get).resource);
        const object = await (0, resourceHandlers_1.resourceCreate)(this._client, path_.toString(), klass, __classPrivateFieldGet(this, _ManagerMixin_handlers, "f"), __classPrivateFieldGet(this, _ManagerMixin_instances, "a", _ManagerMixin_originatingClass_get).methods, innerArgs, idempotencyKey === null || idempotencyKey === void 0 ? void 0 : idempotencyKey.toString());
        return this.postCreateHandler(object, innerArgs);
    }
    async _update(identifier, args) {
        const innerArgs = args || {};
        const { path_ } = innerArgs;
        const klass = await (0, utils_1.getResourceClass)(__classPrivateFieldGet(this, _ManagerMixin_instances, "a", _ManagerMixin_originatingClass_get).resource);
        const object = await (0, resourceHandlers_1.resourceUpdate)(this._client, this.buildPath(innerArgs), identifier, klass, __classPrivateFieldGet(this, _ManagerMixin_handlers, "f"), __classPrivateFieldGet(this, _ManagerMixin_instances, "a", _ManagerMixin_originatingClass_get).methods, innerArgs, path_ === null || path_ === void 0 ? void 0 : path_.toString());
        return this.postUpdateHandler(object, identifier, innerArgs);
    }
    async _delete(identifier, args) {
        const innerArgs = args || {};
        await (0, resourceHandlers_1.resourceDelete)(this._client, this.buildPath(innerArgs), identifier, __classPrivateFieldGet(this, _ManagerMixin_handlers, "f"));
        return this.postDeleteHandler(identifier, innerArgs);
    }
    /* eslint-disable class-methods-use-this, @typescript-eslint/no-unused-vars */
    postListHandler(objects, args) {
        return objects;
    }
    postGetHandler(object, identifier, args) {
        return object;
    }
    postCreateHandler(object, args) {
        return object;
    }
    postUpdateHandler(object, identifier, args) {
        return object;
    }
    postDeleteHandler(identifier, args) {
        return identifier;
    }
    buildPath(args) {
        let path = __classPrivateFieldGet(this, _ManagerMixin_path, "f");
        for (const [key, value] of Object.entries(args || {})) {
            path = path.replace(`{${key}}`, value.toString());
        }
        return path;
    }
}
exports.ManagerMixin = ManagerMixin;
_ManagerMixin_path = new WeakMap(), _ManagerMixin_handlers = new WeakMap(), _ManagerMixin_instances = new WeakSet(), _ManagerMixin_originatingClass_get = function _ManagerMixin_originatingClass_get() {
    return this.constructor;
};
__decorate([
    utils_1.canRaiseHTTPError
], ManagerMixin.prototype, "_list", null);
__decorate([
    utils_1.canRaiseHTTPError
], ManagerMixin.prototype, "_get", null);
__decorate([
    utils_1.canRaiseHTTPError
], ManagerMixin.prototype, "_create", null);
__decorate([
    utils_1.canRaiseHTTPError
], ManagerMixin.prototype, "_update", null);
__decorate([
    utils_1.canRaiseHTTPError
], ManagerMixin.prototype, "_delete", null);
