// fintoc/build/main/lib/mixins/resourceMixin.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _ResourceMixin_instances, _ResourceMixin_handlers, _ResourceMixin_methods, _ResourceMixin_path, _ResourceMixin_attributes, _ResourceMixin_originatingClass_get;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceMixin = void 0;
const resourceHandlers_1 = require("../resourceHandlers");
const utils_1 = require("../utils");
class ResourceMixin {
    constructor(client, handlers, methods, path, content) {
        _ResourceMixin_instances.add(this);
        _ResourceMixin_handlers.set(this, void 0);
        _ResourceMixin_methods.set(this, void 0);
        _ResourceMixin_path.set(this, void 0);
        _ResourceMixin_attributes.set(this, void 0);
        this._client = client;
        __classPrivateFieldSet(this, _ResourceMixin_handlers, handlers, "f");
        __classPrivateFieldSet(this, _ResourceMixin_methods, methods, "f");
        __classPrivateFieldSet(this, _ResourceMixin_path, path, "f");
        __classPrivateFieldSet(this, _ResourceMixin_attributes, [], "f");
        Object.entries(content).forEach(([key, value]) => {
            try {
                this[key] = value;
                __classPrivateFieldGet(this, _ResourceMixin_attributes, "f").push(key);
            }
            catch (_a) { } /* eslint-disable-line no-empty */
        });
    }
    static async _build(client, handlers, methods, path, data) {
        const content = {};
        for (const [key, value] of Object.entries(data)) {
            /* eslint-disable no-await-in-loop */
            const rawResource = this.mappings[key] || key;
            if (value !== undefined) {
                if (Array.isArray(value)) {
                    const resource = (0, utils_1.singularize)(rawResource);
                    const element = value.length > 0 ? value[0] : {};
                    const klass = await (0, utils_1.getResourceClass)(resource, element);
                    content[key] = await Promise.all(value.map((x) => (0, utils_1.objetize)(klass, client, x)));
                }
                else {
                    const klass = await (0, utils_1.getResourceClass)(rawResource, value);
                    content[key] = await (0, utils_1.objetize)(klass, client, value);
                }
            }
            /* eslint-enable no-await-in-loop */
        }
        // @ts-ignore: cannot create an instance of an abstract class
        return new this(client, handlers, methods, path, content);
    }
    _useClient() {
        return this._client;
    }
    _updateClient(newClient) {
        this._client = newClient;
    }
    serialize() {
        let serialized = {};
        __classPrivateFieldGet(this, _ResourceMixin_attributes, "f").forEach((attribute) => {
            const element = (Array.isArray(this[attribute])
                ? this[attribute].map(utils_1.serialize)
                : (0, utils_1.serialize)(this[attribute]));
            serialized = { ...serialized, [attribute]: element };
        });
        return serialized;
    }
    /**
     * Update this instance of the resource.
     *
     * @param args - Data to be passed for the object to be updated with, as specified by the API
     * @returns The updated instance of the object
     */
    update(args) {
        if (!__classPrivateFieldGet(this, _ResourceMixin_methods, "f").includes('update')) {
            throw new TypeError(`${__classPrivateFieldGet(this, _ResourceMixin_instances, "a", _ResourceMixin_originatingClass_get).name}.update is not a valid function of of this resource`);
        }
        return this._update(args);
    }
    /**
     * Delete this instance of the resource.
     *
     * @param args - Object with the arguments to filter the query, using the API parameters
     * @returns The identifier of the deleted object
     */
    delete(args) {
        if (!__classPrivateFieldGet(this, _ResourceMixin_methods, "f").includes('delete')) {
            throw new TypeError(`${__classPrivateFieldGet(this, _ResourceMixin_instances, "a", _ResourceMixin_originatingClass_get).name}.delete is not a valid function of of this resource`);
        }
        return this._delete(args);
    }
    async _update(args) {
        const innerArgs = args || {};
        const id = this[__classPrivateFieldGet(this, _ResourceMixin_instances, "a", _ResourceMixin_originatingClass_get).resourceIdentifier];
        let object = await (0, resourceHandlers_1.resourceUpdate)(this._client, __classPrivateFieldGet(this, _ResourceMixin_path, "f"), id, this.constructor, __classPrivateFieldGet(this, _ResourceMixin_handlers, "f"), __classPrivateFieldGet(this, _ResourceMixin_methods, "f"), innerArgs);
        object = await __classPrivateFieldGet(this, _ResourceMixin_handlers, "f").update(object, id, innerArgs);
        Object.assign(this, object);
        return object;
    }
    async _delete(args) {
        const innerArgs = args || {};
        const identifier = this[__classPrivateFieldGet(this, _ResourceMixin_instances, "a", _ResourceMixin_originatingClass_get).resourceIdentifier];
        await (0, resourceHandlers_1.resourceDelete)(this._client, __classPrivateFieldGet(this, _ResourceMixin_path, "f"), identifier, innerArgs);
        return __classPrivateFieldGet(this, _ResourceMixin_handlers, "f").delete(identifier, innerArgs);
    }
}
exports.ResourceMixin = ResourceMixin;
_ResourceMixin_handlers = new WeakMap(), _ResourceMixin_methods = new WeakMap(), _ResourceMixin_path = new WeakMap(), _ResourceMixin_attributes = new WeakMap(), _ResourceMixin_instances = new WeakSet(), _ResourceMixin_originatingClass_get = function _ResourceMixin_originatingClass_get() {
    return this.constructor;
};
ResourceMixin.mappings = {};
ResourceMixin.resourceIdentifier = 'id';
__decorate([
    utils_1.canRaiseHTTPError
], ResourceMixin.prototype, "_update", null);
__decorate([
    utils_1.canRaiseHTTPError
], ResourceMixin.prototype, "_delete", null);
